

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>ARGs as tree sequences &#8212; Tree Sequence Tutorials</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/tskit-book-theme.css?digest=843e051bda38cea0b82cdbe0f4a2038796abf3af" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/tskit-book-theme.js?digest=6c16a51f00e73652d8545b3e0a23483d12d6db7d"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'args';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tree sequences and simulation" href="simulation_overview.html" />
    <link rel="prev" title="Working with Metadata" href="metadata.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/tskit_logo.svg" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/tskit_logo.svg" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome!
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="what_is.html">What is a tree sequence?</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tree sequence basics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="terminology_and_concepts.html">Terminology &amp; concepts</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started with <strong class="program">tskit</strong></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistics and Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="analysing_tree_sequences.html"><em>Analysing tree sequences</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="analysing_trees.html">Analysing trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="counting_topologies.html"><em>Counting topologies</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="parallelization.html"><em>Parallelization</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Further tskit tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tables_and_editing.html">Tables and editing</a></li>
<li class="toctree-l1"><a class="reference internal" href="simplification.html"><em>Simplification</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Working with Metadata</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#"><em>ARGs as tree sequences</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="simulation_overview.html">Tree sequences and simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="no_mutations.html">Do you really need mutations?</a></li>
<li class="toctree-l1"><a class="reference internal" href="completing_forward_sims.html">Completing forwards simulations</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="advanced_msprime.html">Advanced <strong class="program">msprime</strong> topics</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="demography.html">Demography</a></li>
<li class="toctree-l2"><a class="reference internal" href="bottlenecks.html">Instantaneous Bottlenecks</a></li>
<li class="toctree-l2"><a class="reference internal" href="introgression.html">Introgression</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="forward_sims.html"><em>Building a forward simulator</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Other languages</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tskitr.html">Tskit and R</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tutorial_development.html"><em>Developing new tutorials</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tskit for ...</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="popgen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for population genetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for phylogenetics</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/tskit-dev/tutorials" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tutorials/issues/new?title=Issue%20on%20page%20%2Fargs.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/args.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/tskit-dev/tutorials" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tutorials/issues/new?title=Issue%20on%20page%20%2Fargs.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>ARGs as tree sequences</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#background-semantics">Background semantics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#arg-nodes">ARG nodes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unary-tree-nodes">Unary tree nodes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-likelihoods">Calculating likelihoods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recording-all-nodes-is-expensive">Recording all nodes is expensive</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-the-tree-sequence-graph">Working with the tree sequence graph</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-traversal">Graph traversal</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#traversing-parent-nodes">Traversing parent nodes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#traversing-child-nodes">Traversing child nodes</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#misc">Misc</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#other-software">Other software</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="args-as-tree-sequences">
<span id="sec-args"></span><h1><em>ARGs as tree sequences</em><a class="headerlink" href="#args-as-tree-sequences" title="Permalink to this heading">#</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This tutorial is a work in progress.</p>
</div>
<p>At its heart, a <code class="docutils literal notranslate"><span class="pre">tskit</span></code> <a class="reference internal" href="what_is.html#sec-what-is"><span class="std std-ref">tree sequence</span></a> consists of a list of
<a class="reference internal" href="terminology_and_concepts.html#sec-terminology-nodes"><span class="std std-ref">Nodes</span></a>, and a list of <a class="reference internal" href="terminology_and_concepts.html#sec-terminology-edges"><span class="std std-ref">Edges</span></a> that connect
those nodes. Therefore a succinct tree sequence is equivalent to a
<a class="reference external" href="https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)">mathematical graph</a>,
which is additionally annotated with genomic positions such that at each
position, a path through the edges exists which defines a tree. This graph
interpretation of a tree sequence is tightly connected to the concept of
an “ancestral recombination graph” (or ARG).</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<p>An original, narrower definition, which we do not use here, restricts the term ARG to
the neutral coalescent process with crossover-recombination, and the graph structure
defined by that process, see e.g.
<a class="reference external" href="https://research.monash.edu/en/publications/an-ancestral-recombination-graph">Griffiths &amp; Marjoram (1997)</a></p>
</aside>
<p>The term “ARG” is <a class="reference external" href="https://doi.org/10.1086%2F508901">often used</a> to refer to
a structure consisting of nodes and edges that describe the genetic genealogy of a set
of sampled chromosomes which have evolved via a process of genetic inheritance combined
with recombination. ARGs may contain not just nodes corresponding to genetic
coalescence, but also additional nodes that correspond e.g. to recombination events
(see <a class="reference internal" href="#sec-args-arg-nodes"><span class="std std-ref">ARG nodes</span></a>, below). We call these “full ARGs”, and this tutorial aims
to show you how <a class="reference external" href="https://tskit.dev">tskit</a> can be used to store and analyse them.
As an example, we will generate a full ARG using the process
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_ancestry" title="(in Python)"><code class="docutils literal notranslate"><span class="pre">msprime.sim_ancestry()</span></code></a> with the <code class="docutils literal notranslate"><span class="pre">record_full_arg=True</span></code> option, as described
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-full-arg" title="(in Python)"><span class="xref std std-ref">in the msprime docs</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">parameters</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;samples&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="c1"># Three diploid individuals == six sample genomes</span>
    <span class="s2">&quot;sequence_length&quot;</span><span class="p">:</span> <span class="mf">1e4</span><span class="p">,</span>
    <span class="s2">&quot;recombination_rate&quot;</span><span class="p">:</span> <span class="mf">1e-7</span><span class="p">,</span>
    <span class="s2">&quot;population_size&quot;</span><span class="p">:</span> <span class="mf">1e3</span><span class="p">,</span>
    <span class="s2">&quot;random_seed&quot;</span><span class="p">:</span> <span class="mi">333</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">ts_arg</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="o">**</span><span class="n">parameters</span><span class="p">,</span> <span class="n">record_full_arg</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># NB: the strict Hudson ARG needs unique crossover positions (i.e. a continuous genome)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&quot;Full ARG&quot; simulated under the Hudson model:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot; stored in a tree sequence with </span><span class="si">{</span><span class="n">ts_arg</span><span class="o">.</span><span class="n">num_nodes</span><span class="si">}</span><span class="s2"> nodes and &quot;</span>
    <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">ts_arg</span><span class="o">.</span><span class="n">num_edges</span><span class="si">}</span><span class="s2"> edges which form </span><span class="si">{</span><span class="n">ts_arg</span><span class="o">.</span><span class="n">num_trees</span><span class="si">}</span><span class="s2"> local trees&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&quot;Full ARG&quot; simulated under the Hudson model:
 stored in a tree sequence with 17 nodes and  18 edges which form 3 local trees
</pre></div>
</div>
</div>
</div>
<p>Like any tree sequence, we can also add mutations to the ARG to generate genetic
variation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">mu</span> <span class="o">=</span> <span class="mf">1e-7</span>
<span class="n">ts_arg</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts_arg</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">mu</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">888</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;     Sample node:  &quot;</span> <span class="o">+</span> <span class="s2">&quot;   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ts_arg</span><span class="o">.</span><span class="n">samples</span><span class="p">()))</span>
<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts_arg</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variable site </span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">alleles</span><span class="p">)[</span><span class="n">v</span><span class="o">.</span><span class="n">genotypes</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     Sample node:  0   1   2   3   4   5
Variable site 0: [&#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;C&#39; &#39;C&#39; &#39;C&#39;]
Variable site 1: [&#39;C&#39; &#39;C&#39; &#39;C&#39; &#39;C&#39; &#39;G&#39; &#39;C&#39;]
Variable site 2: [&#39;T&#39; &#39;T&#39; &#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
</pre></div>
</div>
</div>
</div>
<p>As well as the standard <a class="reference internal" href="viz.html#sec-tskit-viz"><span class="std std-ref">Visualization</span></a> of the tree sequence as a set of local
trees, we can also
<a class="reference internal" href="viz.html#sec-tskit-viz-other-graph"><span class="std std-ref">plot this tree sequence in network form</span></a>:</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Incorporate something into <a class="reference external" href="https://github.com/tskit-dev/tsviz">tsviz</a>
like the <code class="docutils literal notranslate"><span class="pre">draw</span></code> function from
https://github.com/tskit-dev/what-is-an-arg-paper/blob/main/argutils/viz.py, and
use that to plot a graph-based viz of this ts, e.g.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">argutils</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ts2</span> <span class="o">=</span> <span class="n">argutils</span><span class="o">.</span><span class="n">viz</span><span class="o">.</span><span class="n">label_nodes</span><span class="p">(</span><span class="n">ts_arg</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">argutils</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">ts_arg</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">draw_edge_widths</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="Example ARG view" src="_images/ARG.png" /></p>
<p>(this PNG file can be removed once the code to autogenerate it is incorporated)</p>
</div>
<section id="background-semantics">
<h2>Background semantics<a class="headerlink" href="#background-semantics" title="Permalink to this heading">#</a></h2>
<p>Two features distinguish the genealogical structure stored in a <code class="docutils literal notranslate"><span class="pre">tskit</span></code> tree sequence
from many other ARG formats. Firstly, the annotations that define which genomic regions
are inherited are stored on <em>edges</em> (via the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Edge.left" title="(in Python)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">left</span></code></a> and a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Edge.right" title="(in Python)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">right</span></code></a>
properties), rather than on the graph nodes, as is sometimes the case. Secondly, the
nodes in a tree sequence correspond to <em>genomes</em>, rather that specific events such as
coalescence or recombination.</p>
<p>Technically therefore, ARGs stored by <code class="docutils literal notranslate"><span class="pre">tskit</span></code> are edge-annotated
“genome ARGs” (gARGs). This results in a flexible format that can describe
ancestral graphs created by many biological processes, including ones that deviate from
the neutral coalescent-with-recombination (CwR), for example ancestries incorporating
gene conversion, or that have evolved under a Wright-Fisher model of inheritance,
in which parents can have more than two children, and coalescence and recombination
can occur in the same generation. The focus on genomes rather than events also
makes it possible to accurately encode ancestry without having to pin down exactly when
the relevant ancestral events took place (TODO: cite our ARG paper).</p>
</section>
<section id="arg-nodes">
<span id="sec-args-arg-nodes"></span><h2>ARG nodes<a class="headerlink" href="#arg-nodes" title="Permalink to this heading">#</a></h2>
<p><a class="reference internal" href="simplification.html#sec-simplification"><span class="std std-ref">Simplified</span></a> tree sequences, such those normally produced by
<code class="docutils literal notranslate"><span class="pre">msprime</span></code>, can be though of as a “simplified ARGs” that contain only nodes that
correspond to a coalescence somewhere in the genome. They are sufficient to capture the
structure of local trees and the correlations between them; this is usually all that is
needed for analysis. However they do not contain complete information about the timings
and topological operations associated with recombination events. This extra information
can be useful for a few specific purposes:</p>
<ol class="arabic simple">
<li><p>Assuming each recombination happens at a unique position, precise information about
which lineages are involved in recombination allows you to work out the exact
tree editing, or <a class="reference internal" href="terminology_and_concepts.html#sec-concepts-sprs"><span class="std std-ref">subtree-prune-and-regraft (SPR)</span></a> moves
required to change one local tree into another as you move along the genome.</p></li>
<li><p>Information about recombination and common ancestor events can be used to calculate
the likelihood of an full ARG under a specific model of evolution (most commonly, the
neutral coalescent with recombination, or CwR, as modelled e.g. by
<a class="reference external" href="https://doi.org/10.1016/0040-5809(83)90013-8">Hudson (1983)</a>)</p></li>
</ol>
<p>Note, however, that it can be impossible to infer non-coalescent
nodes from genetic variation data with any degree of precision.</p>
<section id="unary-tree-nodes">
<span id="sec-args-unary-nodes"></span><h3>Unary tree nodes<a class="headerlink" href="#unary-tree-nodes" title="Permalink to this heading">#</a></h3>
<p>To store additional information about non-coalescent nodes, a full ARG stored in tree
sequence form contains extra <em>unary</em> nodes (i.e. nodes with only one child).
In particular, it can contain <em>recombination nodes</em> which record the timing of
recombination events, and <em>non-coalescent-common-ancestor</em> nodes which record cases
where lineages share a common ancestor but in which genetic material does not coalesce.</p>
<p>The example we have been using is small, and contains just 2 recombination events
(associated with 2 breakpoints). In this instance the only extra nodes happen to be
recombination nodes. <code class="docutils literal notranslate"><span class="pre">Msprime</span></code> only simulates full ARGs in which a recombination event
results in a single crossover, and it records this by storing the two genomes
immediately prior to gamete formation (the genomes that come together to form a
recombinant). In other words, <em>two</em> extra nodes are created for each
recombination: one that captures transmission to the left of the crossover and another,
at an identical time, to the right. These are identified by the
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.msprime.NODE_IS_RE_EVENT" title="(in Python)"><code class="xref py py-data docutils literal notranslate"><span class="pre">NODE_IS_RE_EVENT</span></code></a> flag, and are are highlighed in red
below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the recombination nodes in red, with a horizontal line at the time of occurrence,</span>
<span class="c1"># and only label nodes that are samples or recombination nodes.</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">ts_arg</span><span class="o">.</span><span class="n">samples</span><span class="p">())</span>
<span class="n">re_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">ts_arg</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span> <span class="k">if</span> <span class="n">nd</span><span class="o">.</span><span class="n">flags</span> <span class="o">&amp;</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NODE_IS_RE_EVENT</span><span class="p">)</span>
<span class="n">re_times</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">nd</span><span class="o">.</span><span class="n">time</span><span class="p">)</span> <span class="k">for</span> <span class="n">nd</span> <span class="ow">in</span> <span class="n">ts_arg</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span> <span class="k">if</span> <span class="n">nd</span><span class="o">.</span><span class="n">flags</span> <span class="o">&amp;</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NODE_IS_RE_EVENT</span><span class="p">]</span>
<span class="n">style</span> <span class="o">=</span> <span class="s2">&quot;.y-axis .grid {stroke: #ff000033} .mut .sym {stroke: goldenrod}&quot;</span>
<span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">re_nodes</span><span class="p">:</span>
    <span class="n">style</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;.n</span><span class="si">{</span><span class="n">u</span><span class="si">}</span><span class="s2"> &gt; .sym </span><span class="se">{{</span><span class="s2">fill: red</span><span class="se">}}</span><span class="s2">&quot;</span>
<span class="n">ts_arg</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span>
    <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">y_ticks</span><span class="o">=</span><span class="n">re_times</span><span class="p">,</span>
    <span class="n">y_gridlines</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">style</span><span class="o">=</span><span class="n">style</span><span class="p">,</span>
    <span class="n">mutation_labels</span><span class="o">=</span><span class="p">{},</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="p">{</span><span class="n">u</span><span class="p">:</span> <span class="n">u</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">samples</span> <span class="o">|</span> <span class="n">re_nodes</span><span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7d32c5bd83a271345809a55ca9b50a7e00392e6d29c6e8d5f62459d9955bfe67.svg" src="_images/7d32c5bd83a271345809a55ca9b50a7e00392e6d29c6e8d5f62459d9955bfe67.svg" /></div>
</div>
<p>The location of the recombination nodes imply that the <em>recombination events</em>
happened ~588 and ~59 generations ago. The older one, at 588 generations, involved node
13 (to the left of position 2601.01) and node 14 (to the right). As well as narrowing
down the recombination event to a specific point in time, the position of these two
nodes tells us that the SPR to convert the first into the second tree
involves pruning the branch above samples 1, 3, 4, and 5 and regrafting it onto the
branch above samples 0 and 2, rather than the other way around. Note that this
particular recombination does not change the <em>topology</em> of the tree, but simply the
branch lengths.</p>
<p>The recombination event 59 generations ago involved nodes 7 and 8, with the crossover
ocurring at position 6516.94. The SPR operation which converts the middle tree into the
last one involves pruning the branch above sample node 5 and regrafting it onto the
branch above the common ancestor of 1 and 3. In this case, the recombination has led to
a change in topology, such that the closest relative of 5 is node 4 from positions 0
to 6516.94, but 1 and 3 from positions 6516.94 to 10,000.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Many ARG representations associate each recombination event with a single node rather
than two. It is possible to represent this in <code class="docutils literal notranslate"><span class="pre">tskit</span></code>, but in such an ARG, the
edge annotations do not contain enough information to calculate the standard
likelihood under the Hudson model (see <a class="reference internal" href="#sec-args-likelihoods"><span class="std std-ref">Calculating likelihoods</span></a>).</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Explain in plain language why 2 RE nodes are needed to calculate the likelihood under the
Hudson CwR: see e.g. https://github.com/tskit-dev/msprime/issues/1942#issuecomment-1013718650</p>
<p>One suggested way to do this is to show how there is not enough information in a 1-RE-node
plot to fully recreate the 2-RE-node equivalent. I think this is because we lose
information about the order of breakpoints when multiple breakpoints occur in
the same region of hidden material.</p>
<p>Note also that this approach only applies to a model in which a single crossover occurs
per chromosome.</p>
</div>
</div>
</section>
</section>
<section id="calculating-likelihoods">
<span id="sec-args-likelihoods"></span><h2>Calculating likelihoods<a class="headerlink" href="#calculating-likelihoods" title="Permalink to this heading">#</a></h2>
<p>Because the ARG above was generated under the standard Hudson model (e.g. neutral
evolution in a large population with unique recombination breakpoints along a continuous
genome), we can calculate its likelihood under that model, for a given recombination
rate and population size, using the <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.log_arg_likelihood" title="(in Python)"><code class="docutils literal notranslate"><span class="pre">msprime.log_arg_likelihood()</span></code></a> method.
Note however, that the simulation was run with the default ploidy level of 2, so that the
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_ancestry" title="(in Python)"><code class="docutils literal notranslate"><span class="pre">msprime.sim_ancestry()</span></code></a> method assumed the <code class="docutils literal notranslate"><span class="pre">population_size</span></code> parameter was
the <em>diploid</em> population size. The <code class="docutils literal notranslate"><span class="pre">log_arg_likelihood</span></code> method requires <code class="docutils literal notranslate"><span class="pre">Ne</span></code>, the haploid
population size, which is twice as large, so the likelihood is calculated as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Log likelihood of the genealogy under the Hudson model:&quot;</span><span class="p">,</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">log_arg_likelihood</span><span class="p">(</span>
        <span class="n">ts_arg</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">parameters</span><span class="p">[</span><span class="s2">&quot;recombination_rate&quot;</span><span class="p">],</span>
        <span class="n">Ne</span><span class="o">=</span><span class="n">parameters</span><span class="p">[</span><span class="s2">&quot;population_size&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span>  <span class="c1"># Number of *haploid* genomes</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Log likelihood of the genealogy under the Hudson model: -93.57791165409245
</pre></div>
</div>
</div>
</div>
<p>It is worth noting that we fully simplify the tree above, we remove all the unary nodes
and therefore lose information about the timings of recombination and non-coalescent
common ancestry, but we still keep the local trees intact:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">ts_arg</span><span class="o">.</span><span class="n">simplify</span><span class="p">()</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span>
    <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="p">{</span><span class="n">u</span><span class="p">:</span> <span class="n">u</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()},</span>
    <span class="n">mutation_labels</span><span class="o">=</span><span class="p">{},</span>
    <span class="n">style</span><span class="o">=</span><span class="s2">&quot;.mut .sym {stroke: goldenrod}&quot;</span><span class="p">,</span>
    <span class="n">y_ticks</span><span class="o">=</span><span class="p">[</span><span class="n">t</span><span class="o">*</span><span class="mi">500</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">)]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ccd8b909f77ca1a5c77922f86baf2bd128eab494040c6a69549555754131ee4f.svg" src="_images/ccd8b909f77ca1a5c77922f86baf2bd128eab494040c6a69549555754131ee4f.svg" /></div>
</div>
<p>Because of this loss of information, the ARG likelihood cannot be calculated from the
simplified tree sequence. We can still, however, calculate the <em>mutation likelihood</em>
(i.e. the likelihood of the observed pattern of mutations, given the genealogy) because
the topology and branch lengths of the local trees remain unchanged after simplification:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Log likelihood of mutations given the genealogy:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39; &quot;full&quot; ARG:&#39;</span><span class="p">,</span>  <span class="n">msprime</span><span class="o">.</span><span class="n">log_mutation_likelihood</span><span class="p">(</span><span class="n">ts_arg</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mu</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot; simplified:&quot;</span><span class="p">,</span> <span class="n">msprime</span><span class="o">.</span><span class="n">log_mutation_likelihood</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mu</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Log likelihood of mutations given the genealogy:
 &quot;full&quot; ARG: -33.204860594400216
 simplified: -33.204860594400216
</pre></div>
</div>
</div>
</div>
</section>
<section id="recording-all-nodes-is-expensive">
<h2>Recording all nodes is expensive<a class="headerlink" href="#recording-all-nodes-is-expensive" title="Permalink to this heading">#</a></h2>
<p>Many extra nodes are required to store full information about ancestrally relevant
recombination. In fact, as the sequence length increases, these non-coalescent nodes come
to dominate the tree sequence (which is one reason they are not included by default).
We can calculate the percentage of non-coalescent nodes by comparing a full ARG with
its simplified version:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">large_sim_parameters</span> <span class="o">=</span> <span class="n">parameters</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">large_sim_parameters</span><span class="p">[</span><span class="s2">&quot;sequence_length&quot;</span><span class="p">]</span> <span class="o">*=</span> <span class="mi">1000</span>
<span class="n">large_ts_arg</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="o">**</span><span class="n">large_sim_parameters</span><span class="p">,</span> <span class="n">record_full_arg</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">large_ts</span> <span class="o">=</span> <span class="n">large_ts_arg</span><span class="o">.</span><span class="n">simplify</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Non-coalescent nodes take up &quot;</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">large_ts</span><span class="o">.</span><span class="n">num_nodes</span><span class="o">/</span><span class="n">large_ts_arg</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="si">:</span><span class="s2">0.2f</span><span class="si">}</span><span class="s2">% &quot;</span>
    <span class="sa">f</span><span class="s2">&quot;of this </span><span class="si">{</span><span class="n">large_ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="o">/</span><span class="mf">1e6</span><span class="si">:</span><span class="s2">g</span><span class="si">}</span><span class="s2"> megabase </span><span class="si">{</span><span class="n">large_ts</span><span class="o">.</span><span class="n">num_samples</span><span class="si">}</span><span class="s2">-tip ARG&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Non-coalescent nodes take up 99.32% of this 10 megabase 6-tip ARG
</pre></div>
</div>
</div>
</div>
<p>This is one of the primary reasons that nodes which are never associated with coalescent
events are excluded by default in simulation software such as
<a class="reference external" href="https://tskit.dev/software/msprime.html">msprime</a> and
<a class="reference external" href="https://tskit.dev/software/SLiM.html">SLiM</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As well as ancestrally relevant nodes, the original (mathematical) ARG formulation by
<a class="reference external" href="https://www.jstor.org/stable/4355649">Griffiths (1991)</a> includes recombination
nodes that are not ancestral to the samples. This leads to a graph with an
vastly larger number of nodes than even the ARGs simulated here, and using such
structures for simulation or inference is therefore infeasible.</p>
</div>
</section>
<section id="working-with-the-tree-sequence-graph">
<h2>Working with the tree sequence graph<a class="headerlink" href="#working-with-the-tree-sequence-graph" title="Permalink to this heading">#</a></h2>
<p>All tree sequences, including, but not limited to full ARGs, can be treated as
directed (acyclic) graphs. Although many tree sequence operations operate from left to
right along the genome, some are more naturally though of as passing from node
to node via the edges, regardless of the genomic position of the edge. This section
describes some of these fundamental graph operations.</p>
<section id="graph-traversal">
<h3>Graph traversal<a class="headerlink" href="#graph-traversal" title="Permalink to this heading">#</a></h3>
<p>The standard edge iterator, <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.edge_diffs" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.edge_diffs()</span></code></a>, goes from left to
right along the genome, matching the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.trees" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> iterator. Although
this will visit all the edges in the graph, these will <em>not</em> necessarily be grouped
by the node ID either of the edge parent or the edge child. To do this, an alternative
traversal (from top-to-bottom or bottom-to-top of the tree sequence) is required.</p>
<p>To traverse the graph by node ID, the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.nodes" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a> iterator can be
used. In particular, because parents are required to be strictly older than their
children, iterating through nodes using <code class="docutils literal notranslate"><span class="pre">order=&quot;timeasc&quot;</span></code> will ensure that children
are always visited before their parents (similar to a breadth-first or “level order”
search). However, using <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.nodes" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a> is inefficient if you also
want to access the <em>edges</em> associated with each node.</p>
<p>The examples below show how to efficiently visit the all the edges in a
tree sequence, grouped by the nodes to which they are connected, while
also ensuring that children are visited before parents (or vice versa).</p>
<section id="traversing-parent-nodes">
<h4>Traversing parent nodes<a class="headerlink" href="#traversing-parent-nodes" title="Permalink to this heading">#</a></h4>
<p>The most efficient graph traversal method visits all the parent nodes in the
tree sequence, grouping edges for which that node is a parent. This is simple
because edges in a tree sequence are <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-requirements" title="(in Python)"><span class="xref std std-ref">ordered</span></a>
firstly by the time of the parent node, then by node ID.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">operator</span>
<span class="k">def</span> <span class="nf">edges_by_parent_timeasc</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">edges</span><span class="p">(),</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;parent&quot;</span><span class="p">))</span>

<span class="k">for</span> <span class="n">parent_id</span><span class="p">,</span> <span class="n">edges</span> <span class="ow">in</span> <span class="n">edges_by_parent_timeasc</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">parent_id</span><span class="p">)</span><span class="o">.</span><span class="n">time</span>
    <span class="n">children</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="o">.</span><span class="n">child</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">edges</span><span class="p">}</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Node </span><span class="si">{</span><span class="n">parent_id</span><span class="si">}</span><span class="s2"> at time </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> has these child node IDs: </span><span class="si">{</span><span class="n">children</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node 6 at time 56.0454802286223 has these child node IDs: {1, 3}
Node 7 at time 120.7850821708138 has these child node IDs: {5, 6}
Node 8 at time 142.86643656492956 has these child node IDs: {4, 5}
Node 9 at time 236.7237255340037 has these child node IDs: {0, 2}
Node 10 at time 272.05118690557765 has these child node IDs: {8, 4, 6, 7}
Node 11 at time 610.912738605865 has these child node IDs: {9, 10}
Node 12 at time 1382.4658244607965 has these child node IDs: {9, 10}
</pre></div>
</div>
</div>
</div>
<p>This visits children before parents. To visit parents before children, you can
simply use <code class="docutils literal notranslate"><span class="pre">reversed(ts.edges())</span></code> rather than <code class="docutils literal notranslate"><span class="pre">ts.edges()</span></code> within the
<code class="docutils literal notranslate"><span class="pre">groupby</span></code> function. Note that terminal nodes (i.e. which are either isolated
or leaf nodes in all local trees) are not visited by this function: therefore
the method above will omit all the terminal sample nodes.</p>
</section>
<section id="traversing-child-nodes">
<h4>Traversing child nodes<a class="headerlink" href="#traversing-child-nodes" title="Permalink to this heading">#</a></h4>
<p>Sometimes you may wish to iterate over all the edges for which a node is a
child (rather than a parent). This can be done by sorting the edges by
child node time (then child node id). This is a little slower, but can be
done relatively efficiently as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">operator</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="k">def</span> <span class="nf">edges_by_child_timeasc</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="c1"># edges sorted by child node time, then child node id using np.lexsort</span>
    <span class="n">it</span> <span class="o">=</span> <span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">lexsort</span><span class="p">((</span><span class="n">ts</span><span class="o">.</span><span class="n">edges_child</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes_time</span><span class="p">[</span><span class="n">ts</span><span class="o">.</span><span class="n">edges_child</span><span class="p">])))</span>
    <span class="k">return</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">it</span><span class="p">,</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s2">&quot;child&quot;</span><span class="p">))</span>

<span class="k">for</span> <span class="n">child_id</span><span class="p">,</span> <span class="n">edges</span> <span class="ow">in</span> <span class="n">edges_by_child_timeasc</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">child_id</span><span class="p">)</span><span class="o">.</span><span class="n">time</span>
    <span class="n">parents</span> <span class="o">=</span> <span class="p">{</span><span class="n">e</span><span class="o">.</span><span class="n">parent</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">edges</span><span class="p">}</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Node </span><span class="si">{</span><span class="n">child_id</span><span class="si">}</span><span class="s2"> at time </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> has these parent node IDs: </span><span class="si">{</span><span class="n">parents</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node 0 at time 0.0 has these parent node IDs: {9}
Node 1 at time 0.0 has these parent node IDs: {6}
Node 2 at time 0.0 has these parent node IDs: {9}
Node 3 at time 0.0 has these parent node IDs: {6}
Node 4 at time 0.0 has these parent node IDs: {8, 10}
Node 5 at time 0.0 has these parent node IDs: {8, 7}
Node 6 at time 56.0454802286223 has these parent node IDs: {10, 7}
Node 7 at time 120.7850821708138 has these parent node IDs: {10}
Node 8 at time 142.86643656492956 has these parent node IDs: {10}
Node 9 at time 236.7237255340037 has these parent node IDs: {11, 12}
Node 10 at time 272.05118690557765 has these parent node IDs: {11, 12}
</pre></div>
</div>
</div>
</div>
<p>To visit parents before children, the <code class="docutils literal notranslate"><span class="pre">lexsort</span></code> can take the negated
<code class="docutils literal notranslate"><span class="pre">-ts.nodes_time</span></code> rather than simply using <code class="docutils literal notranslate"><span class="pre">ts.nodes_time</span></code>. Note that nodes
which are never children of an edge are not visited by this algorithm. Such
nodes are either <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-data-model-tree-isolated-nodes" title="(in Python)"><span class="xref std std-ref">isolated</span></a> or a
{ref}`root&lt;sec_data_model_tree_roots&gt; in each local tree.</p>
</section>
</section>
<section id="misc">
<h3>Misc<a class="headerlink" href="#misc" title="Permalink to this heading">#</a></h3>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Add extra content as per <a class="github reference external" href="https://github.com/tskit-dev/tutorials/issues/43">tskit-dev/tutorials#43</a></p>
</div>
</section>
</section>
<section id="other-software">
<h2>Other software<a class="headerlink" href="#other-software" title="Permalink to this heading">#</a></h2>
<div class="admonition-todo admonition" id="id4">
<p class="admonition-title">Todo</p>
<p>Show how ARGweaver output can be converted to tskit form.</p>
</div>
<div class="admonition-todo admonition" id="id5">
<p class="admonition-title">Todo</p>
<p>Show how KwARG output can be converted to tskit form.</p>
</div>
<div class="admonition-todo admonition" id="id6">
<p class="admonition-title">Todo</p>
<p>Implement conversion between the 2 RE node version and the 1 RE node version</p>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="metadata.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Working with Metadata</p>
      </div>
    </a>
    <a class="right-next"
       href="simulation_overview.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Tree sequences and simulation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#background-semantics">Background semantics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#arg-nodes">ARG nodes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unary-tree-nodes">Unary tree nodes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-likelihoods">Calculating likelihoods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recording-all-nodes-is-expensive">Recording all nodes is expensive</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-the-tree-sequence-graph">Working with the tree sequence graph</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-traversal">Graph traversal</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#traversing-parent-nodes">Traversing parent nodes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#traversing-child-nodes">Traversing child nodes</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#misc">Misc</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#other-software">Other software</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>