
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Building a forward simulator &#8212; Tree Sequence Tutorials</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!-- 
    this give us a css class that will be invisible only if js is disabled 
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=26a4bc78f4c0ddb94549"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549" />

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'forward_sims';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Advanced forward simulations" href="more_forward_sims.html" />
    <link rel="prev" title="Completing forwards simulations" href="completing_forward_sims.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tskit_logo.svg" class="logo__image only-light" alt="Tree Sequence Tutorials - Home"/>
    <img src="_static/tskit_logo.svg" class="logo__image only-dark pst-js-only" alt="Tree Sequence Tutorials - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="what_is.html">What is a tree sequence?</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tree sequence basics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="terminology_and_concepts.html">Terminology &amp; concepts</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started with <strong class="program">tskit</strong></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistics and Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="analysing_tree_sequences.html"><em>Analysing tree sequences</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="analysing_trees.html">Analysing trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="incremental_algorithms.html"><em>Incremental algorithms</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="counting_topologies.html">Counting topologies</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallelization.html"><em>Parallelization</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Further tskit tutorials</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tables_and_editing.html">Tables and editing</a></li>
<li class="toctree-l1"><a class="reference internal" href="simplification.html"><em>Simplification</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Working with Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="args.html">ARGs as tree sequences</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="simulation_overview.html">Tree sequences and simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="no_mutations.html">Do you really need mutations?</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="advanced_msprime.html">Advanced <strong class="program">msprime</strong> topics</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="demography.html">Demography</a></li>
<li class="toctree-l2"><a class="reference internal" href="bottlenecks.html">Instantaneous Bottlenecks</a></li>
<li class="toctree-l2"><a class="reference internal" href="introgression.html">Introgression</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="completing_forward_sims.html">Completing forwards simulations</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Building a forward simulator</a></li>
<li class="toctree-l1"><a class="reference internal" href="more_forward_sims.html"><em>Advanced forward simulations</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Other languages</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tskitr.html">Tskit and R</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tutorial_development.html"><em>Developing new tutorials</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tskit for ...</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="popgen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for population genetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for phylogenetics</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/forward_sims.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Building a forward simulator</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#definitions">Definitions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#approach">Approach</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#core-steps">Core steps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inheritance-without-recombination">Inheritance without recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inheritance-with-recombination">Inheritance with recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-examples">Basic examples</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simplification">Simplification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-tskit-forward-simulations-simplification-id-changes">ID changes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-intermediate-nodes">Removing intermediate nodes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination">Recombination</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#larger-simulations">Larger simulations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ensuring-coalescence">Ensuring coalescence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#more-complex-forward-simulations">More complex forward-simulations</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="building-a-forward-simulator">
<span id="sec-tskit-forward-simulations"></span><h1>Building a forward simulator<a class="headerlink" href="#building-a-forward-simulator" title="Link to this heading">#</a></h1>
<p>This tutorial shows how tskit can be used to build your own forward-time, tree sequence simulator from scratch.
The simulator will use the discrete-time Wright-Fisher model, and track individuals
along with their genomes, storing inherited genomic regions as well as the full pedigree.</p>
<p>The code in this tutorial is broken into separate functions for clarity and
to make it easier to modify for your own purposes; a simpler and substantially
condensed forward-simulator is coded as a single function at the top of the
<a class="reference internal" href="completing_forward_sims.html#sec-completing-forwards-simulations"><span class="std std-ref">Completing forwards simulations</span></a> tutorial.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are simply trying to obtain a tree sequence which is
the result of a forward-time simulation, this can be done by using one of the
highly capable forward-time genetic simulators that already exist, such as
<a class="reference external" href="https://messerlab.org/slim/">SLiM</a> or <a class="reference external" href="https://github.com/molpopgen/fwdpy11">fwdpy11</a>.
Documentation and tutorials for these tools exist on their respective websites. This
tutorial is instead intended to illustrate the general principles that lie behind such
simulators.</p>
</div>
<p>We will focus on the case of diploids, in which each <a class="reference external" href="https://tskit.dev/tskit/docs/stable/glossary.html#sec-nodes-or-individuals" title="(in Project name not set)"><span class="xref std std-ref">individual</span></a> contains 2 genomes,
but the concepts used here generalize to any ploidy, if you are willing to do the book-keeping.
The individuals themselves are not strictly necessary for representing genetic genealogies
(it’s the genomes which are important), but they are needed during the simulation,
and so we record them in the resulting output for completeness.</p>
<section id="definitions">
<h2>Definitions<a class="headerlink" href="#definitions" title="Link to this heading">#</a></h2>
<p>Before we can make any progress, we require a few definitions.</p>
<p>A <em>node</em> represents a genome at a point in time (often we imagine this as the “birth time” of the genome).
It can be described by a tuple, <code class="docutils literal notranslate"><span class="pre">(id,</span> <span class="pre">time)</span></code>, where <code class="docutils literal notranslate"><span class="pre">id</span></code> is a unique integer,
and <code class="docutils literal notranslate"><span class="pre">time</span></code> reflects the birth time of that <code class="docutils literal notranslate"><span class="pre">id</span></code>. When generating a tree sequence,
this will be stored in a row of the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-node-table-definition" title="(in Project name not set)"><span>Node Table</span></a>.</p>
<p>A <em>diploid individual</em> is a group of two nodes. During simulation,
a simple and efficient grouping assigns sequential pairs of node IDs to an individual.
It can be helpful (but not strictly necessary) to store individuals within the tree sequence
as rows of the the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-individual-table-definition" title="(in Project name not set)"><span>Individual Table</span></a> (a node can then be assigned to an
individual by storing that individual’s id in the appropriate row of the node table).</p>
<p>An <em>edge</em> reflects a transmission event between nodes.  An edge is a tuple <code class="docutils literal notranslate"><span class="pre">(Left,</span> <span class="pre">Right,</span> <span class="pre">Parent,</span> <span class="pre">Child)</span></code>
whose meaning is “The <code class="docutils literal notranslate"><span class="pre">Child</span></code> genome inherited the genomic interval [<code class="docutils literal notranslate"><span class="pre">Left</span></code>, <code class="docutils literal notranslate"><span class="pre">Right</span></code>) from the <code class="docutils literal notranslate"><span class="pre">Parent</span></code> genome”.
In a tree sequence this is stored in a row of the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Project name not set)"><span>Edge Table</span></a>.</p>
<p>The <em>time</em>, in the discrete-time Wright-Fisher (WF) model which we will simulate, is measured in
integer generations. To match the tskit notion of time, we record time in <em>generations ago</em>:
i.e. for a simple simulation of G generations, we start the simulation at generation <span class="math notranslate nohighlight">\(G-1\)</span>
and count down until we reach generation 0 (the current-day).</p>
<p>The <em>population</em> consists of <span class="math notranslate nohighlight">\(N\)</span> diploid individuals (<span class="math notranslate nohighlight">\(2N\)</span> nodes) at a particular time <span class="math notranslate nohighlight">\(t\)</span>.
At the start, the population will have no known ancestry, but subsequently
each individual will be formed by choosing (at random) two parent individuals from the
population in the previous generation.</p>
</section>
<section id="approach">
<h2>Approach<a class="headerlink" href="#approach" title="Link to this heading">#</a></h2>
<p>We will generate edges, nodes, and individuals forwards in time, adding them to the relevant <code class="docutils literal notranslate"><span class="pre">tskit</span></code> tables.
To aid efficiency, we will also see how to “simplify” the tables into the minimal set of nodes and edges that
describe the history of the sample. Finally, these tables can be exported into an immutable
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> for storing or analysis.</p>
<section id="setup">
<h3>Setup<a class="headerlink" href="#setup" title="Link to this heading">#</a></h3>
<p>First, we’ll import the necessary libraries and define some general parameters.
The <a class="reference external" href="https://numpy.org/doc/stable/">numpy</a> library will be used to produce random numbers.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">tskit</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">random_seed</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">random</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">random_seed</span><span class="p">)</span>  <span class="c1"># A random number generator for general use</span>

<span class="n">L</span> <span class="o">=</span> <span class="mi">50_000</span>  <span class="c1"># The sequence length: 50 Kb</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="core-steps">
<h3>Core steps<a class="headerlink" href="#core-steps" title="Link to this heading">#</a></h3>
<p>The core simulation function generates a new population by repeating the following steps:</p>
<ol class="arabic simple">
<li><p>Pick a pair of parent individuals at random from the population in the previous generation.</p></li>
<li><p>Create a new child individual with two genomes (an individual ID will be created
by <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTable.add_row" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adding</span> <span class="pre">a</span> <span class="pre">row</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">individual</span> <span class="pre">table</span></code></a>
and two nodes IDs will be created by
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NodeTable.add_row" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adding</span> <span class="pre">two</span> <span class="pre">rows</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">node</span> <span class="pre">table</span></code></a>. If desired,
we can also provide the parent IDs when adding to the individual table, which
will result in storing the entire genealogical pedigree.</p></li>
<li><p>Add the inheritance paths for each child genome using an <code class="docutils literal notranslate"><span class="pre">add_inheritance_paths()</span></code>
function (defined later).</p></li>
</ol>
<p>We will also define a simpler function, involving just step 2, to create the initial
population.</p>
<p>For convenience, the population will be stored in a Python dictionary that maps
the individual ID to the IDs of its two genomes. Note that in lower-level
languages such as C, rather than use a mapping of individual ID to two genomes,
we could instead rely on the fact that genome IDs are allocated sequentially, and
pick a random pair of genomes from the previous population directly,
without using individual IDs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For visualising unsimplified tree sequences, it can help to flag all nodes as samples</span>
<span class="n">default_node_flags</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span>

<span class="k">def</span> <span class="nf">make_diploid</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">parent_individuals</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Make an individual and its diploid genomes by adding to tables, returning the IDs.</span>
<span class="sd">    Specifying parent_individuals is optional but results in the pedigree being stored.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">individual_id</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="n">parent_individuals</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">individual_id</span><span class="p">,</span> <span class="p">(</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">default_node_flags</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">individual</span><span class="o">=</span><span class="n">individual_id</span><span class="p">),</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">default_node_flags</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">individual</span><span class="o">=</span><span class="n">individual_id</span><span class="p">),</span>
    <span class="p">)</span>

<span class="k">def</span> <span class="nf">new_population</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">prev_pop</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
    <span class="n">pop</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># fill with individual_ID: (maternal_genome_ID, paternal_genome_ID)</span>

    <span class="c1"># Cache the list of individual IDs in the previous population, for efficiency</span>
    <span class="n">prev_individuals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">prev_pop</span><span class="o">.</span><span class="n">keys</span><span class="p">()],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">prev_pop</span><span class="p">)):</span>
        <span class="c1"># 1. Pick two individual parent IDs at random, `replace=True` allows selfing</span>
        <span class="n">mother_and_father</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">prev_individuals</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># 2. Get 1 new individual ID + 2 new node IDs</span>
        <span class="n">child_id</span><span class="p">,</span> <span class="n">child_genomes</span> <span class="o">=</span> <span class="n">make_diploid</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">mother_and_father</span><span class="p">)</span>
        <span class="n">pop</span><span class="p">[</span><span class="n">child_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">child_genomes</span>  <span class="c1"># store the genome IDs</span>
        
        <span class="c1"># 3. Add inheritance paths to both child genomes</span>
        <span class="k">for</span> <span class="n">child_genome</span><span class="p">,</span> <span class="n">parent_individual</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">child_genomes</span><span class="p">,</span> <span class="n">mother_and_father</span><span class="p">):</span>
            <span class="n">parent_genomes</span> <span class="o">=</span> <span class="n">prev_pop</span><span class="p">[</span><span class="n">parent_individual</span><span class="p">]</span>
            <span class="n">add_inheritance_paths</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">parent_genomes</span><span class="p">,</span> <span class="n">child_genome</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pop</span>

<span class="k">def</span> <span class="nf">initialise_population</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">size</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
    <span class="c1"># Just return a dictionary by repeating step 2 above</span>
    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">make_diploid</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">time</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For simplicity, the code above assumes any parent can be a mother or a father (i.e. this is a hermaphrodite species).
It also allows the same parent to be chosen as a mother and as a father (i.e. “selfing” is allowed),
which gives simpler theoretical results. This is easy to change if required.</p>
</div>
<p>Our forward-time simulator simply involves repeatedly running the <code class="docutils literal notranslate"><span class="pre">next_population()</span></code> routine,
replacing the old population with the new one. For efficiency reasons, <code class="docutils literal notranslate"><span class="pre">tskit</span></code> has strict requirements
for the order of edges in the edge table, so we need to <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection.sort" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a> the tables before we output the final tree sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">forward_WF</span><span class="p">(</span><span class="n">num_diploids</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">,</span> <span class="n">generations</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run a forward-time Wright Fisher simulation of a diploid population, returning</span>
<span class="sd">    a tree sequence representing the genetic genealogy of the simulated genomes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">global</span> <span class="n">random</span>
    <span class="n">random</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">random_seed</span><span class="p">)</span> 
    <span class="n">tables</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">time_units</span> <span class="o">=</span> <span class="s2">&quot;generations&quot;</span>  <span class="c1"># optional, but helpful when plotting</span>

    <span class="n">pop</span> <span class="o">=</span> <span class="n">initialise_population</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">generations</span><span class="p">,</span> <span class="n">num_diploids</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">generations</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">generations</span> <span class="o">=</span> <span class="n">generations</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">pop</span> <span class="o">=</span> <span class="n">new_population</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">generations</span><span class="p">,</span> <span class="n">pop</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">)</span>

    <span class="n">tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="inheritance-without-recombination">
<h3>Inheritance without recombination<a class="headerlink" href="#inheritance-without-recombination" title="Link to this heading">#</a></h3>
<p>The final piece of the simulation is to define the <code class="docutils literal notranslate"><span class="pre">add_inheritance_paths()</span></code> function,
which saves the inheritance paths in the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Project name not set)"><span>Edge Table</span></a>.
For reference, the simplest case (a small focal region in which there is no recombination)
can be coded as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">add_inheritance_paths</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">parent_nodes</span><span class="p">,</span> <span class="n">child_node</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">):</span>
    <span class="s2">&quot;Add inheritance paths from a randomly chosen parent genome to the child genome.&quot;</span>
    <span class="k">assert</span> <span class="n">recombination_rate</span> <span class="o">==</span> <span class="mi">0</span>
    <span class="n">left</span><span class="p">,</span> <span class="n">right</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20_000</span><span class="p">,</span> <span class="mi">21_000</span><span class="p">]</span>  <span class="c1"># only define inheritance in this focal region</span>
    <span class="n">inherit_from</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  <span class="c1"># randomly choose the 1st or the 2nd parent node</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">,</span> <span class="n">parent_nodes</span><span class="p">[</span><span class="n">inherit_from</span><span class="p">],</span> <span class="n">child_node</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="inheritance-with-recombination">
<h3>Inheritance with recombination<a class="headerlink" href="#inheritance-with-recombination" title="Link to this heading">#</a></h3>
<p>Recombination adds complexity to the inheritance paths from a child to its parents,
that the child inherits a mosaic of the two genomes present in each parent.</p>
<p>The exact details of the mosaic will depend on the model of recombination you
wish to implement. For instance, a simple model such as that in the
<a class="reference internal" href="completing_forward_sims.html#sec-completing-forwards-simulations"><span class="std std-ref">Completing forwards simulations</span></a> tutorial might assume exactly one
crossover per chromosome. A complex model might allow not just multiple
crossovers with e.g. recombination “hotspots”, but also non-crossover events
such as <a class="reference external" href="https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-gene-conversion" title="(in Project name not set)"><span>Gene conversion</span></a>.</p>
<p>Below is a redefined <code class="docutils literal notranslate"><span class="pre">add_inheritance_paths()</span></code> function of intermediate complexity,
which models recombination as a uniform Poisson process along a genomic interval.
We generate a set of “breakpoints” along the genome, then allocates edges from the
child genome to one or the other genome from the parent, with left and right positions
reflecting the breakpoint positions. Note that real recombination rates are usually
such that they result in relatively
few breakpoints per chromosome (in humans, around 1 or 2).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">add_inheritance_paths</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">parent_genomes</span><span class="p">,</span> <span class="n">child_genome</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="p">):</span>
    <span class="s2">&quot;Add paths from parent genomes to the child genome, with crossover recombination.&quot;</span>
    <span class="n">L</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">sequence_length</span>
    <span class="n">num_recombinations</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">recombination_rate</span> <span class="o">*</span> <span class="n">L</span><span class="p">)</span>
    <span class="n">breakpoints</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">num_recombinations</span><span class="p">)</span>
    <span class="n">breakpoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(([</span><span class="mi">0</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">breakpoints</span><span class="p">),</span> <span class="p">[</span><span class="n">L</span><span class="p">]))</span>
    <span class="n">inherit_from</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  <span class="c1"># starting parental genome</span>

    <span class="c1"># iterate over pairs of ([0, b1], [b1, b2], [b2, b3], ... [bN, L])</span>
    <span class="k">for</span> <span class="n">left</span><span class="p">,</span> <span class="n">right</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">breakpoints</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">breakpoints</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span>
            <span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">,</span> <span class="n">parent_genomes</span><span class="p">[</span><span class="n">inherit_from</span><span class="p">],</span> <span class="n">child_genome</span><span class="p">)</span>
        <span class="n">inherit_from</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">inherit_from</span>  <span class="c1"># switch to other parent genome</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Above, breakpoint positions occur on a continuous line (i.e. “infinite breakpoint positions”),
to match population genetic theory. It is relatively easy to alter this to
allow recombinations only at integer positions</p>
</div>
</section>
<section id="basic-examples">
<h3>Basic examples<a class="headerlink" href="#basic-examples" title="Link to this heading">#</a></h3>
<p>Now we can test the <code class="docutils literal notranslate"><span class="pre">forward_WF()</span></code> function for a single generation with a small
population size of 6 diploids, and print out the resulting tree sequence. For simplicity,
we will omit recombination for now.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">forward_WF</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="mi">200</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/57fb682d33632da1951228f29d6547f511d28d3bfdf70520870c817b247decaf.svg" src="_images/57fb682d33632da1951228f29d6547f511d28d3bfdf70520870c817b247decaf.svg" />
</div>
</div>
<p>It looks like it is working correctly: all 12 genomes (6 diploids) in the current generation at time=0 trace back to a
genome in the initial generation at time=1. Note that not all individuals in the initial generation have passed on genetic
material at this genomic position (they appear as isolated nodes at the top of the plot).</p>
<p>Now let’s simulate for a longer time period, and set a few helpful plotting parameters.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>By convention we plot the most recent generation at the bottom of the plot
(i.e. perversely, each “tree” has leaves towards the bottom, and roots at the top)</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">forward_WF</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="n">graphics_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;y_axis&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s2">&quot;y_label&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Time (</span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">time_units</span><span class="si">}</span><span class="s2"> ago)&quot;</span><span class="p">,</span>
    <span class="s2">&quot;y_ticks&quot;</span><span class="p">:</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="s1">&#39;Current&#39;</span> <span class="k">if</span> <span class="n">i</span><span class="o">==</span><span class="mi">0</span> <span class="k">else</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">21</span><span class="p">)},</span>
<span class="p">}</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">1200</span><span class="p">,</span> <span class="mi">400</span><span class="p">),</span> <span class="o">**</span><span class="n">graphics_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3e1420a8906e654a58eddab751c771ec003332228ac76b23c22497b12629f665.svg" src="_images/3e1420a8906e654a58eddab751c771ec003332228ac76b23c22497b12629f665.svg" />
</div>
</div>
<p>This is starting to look like a real genealogy! Clearly, however, there are many
“extinct” lineages that have not made it to the current day.</p>
</section>
</section>
<section id="simplification">
<span id="sec-tskit-forward-simulations-simplification"></span><h2>Simplification<a class="headerlink" href="#simplification" title="Link to this heading">#</a></h2>
<p>The key to efficent forward-time genealogical simulation is the process of <a class="reference internal" href="simplification.html#sec-simplification"><span class="std std-ref">Simplification</span></a>,
which can reduce much of the complexity shown in the tree above.
Typically, we want to remove all the lineages that do not contribute to the current day genomes.
We do this via the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> method, specifying that only the nodes
in the current generation are “samples”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">current_day_genomes</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">current_day_genomes</span><span class="p">,</span> <span class="n">keep_unary</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">filter_nodes</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">simplified_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span> <span class="mi">400</span><span class="p">),</span> <span class="o">**</span><span class="n">graphics_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/b7ba1f1279c9c5b5c6f64ca42862320ebcb857bab6bf5e91891ac225e57f7362.svg" src="_images/b7ba1f1279c9c5b5c6f64ca42862320ebcb857bab6bf5e91891ac225e57f7362.svg" />
</div>
</div>
<section id="sec-tskit-forward-simulations-simplification-id-changes">
<span id="id-changes"></span><h3>ID changes<a class="headerlink" href="#sec-tskit-forward-simulations-simplification-id-changes" title="Link to this heading">#</a></h3>
<p>We just simplified with <code class="docutils literal notranslate"><span class="pre">filter_nodes=False</span></code>, meaning that the tree sequence retained
all nodes even after simplification, even those that are no longer part of
the genealogy. By default (if <code class="docutils literal notranslate"><span class="pre">filter_nodes</span></code> is not specified), these nodes are removed,
which changes the node IDs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">current_day_genomes</span><span class="p">,</span> <span class="n">keep_unary</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">simplified_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span> <span class="o">**</span><span class="n">graphics_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7e3383c6889fe3389855230a4a8aa9471ae75ec92700699da6a1a516f55f19fa.svg" src="_images/7e3383c6889fe3389855230a4a8aa9471ae75ec92700699da6a1a516f55f19fa.svg" />
</div>
</div>
<p>You can see that the list of nodes passed to <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> (i.e. the current-day genomes)
have become the first nodes in the table, numbered from 0..11;
the remaining (internal) nodes have been renumbered from youngest to oldest.</p>
</section>
<section id="removing-intermediate-nodes">
<h3>Removing intermediate nodes<a class="headerlink" href="#removing-intermediate-nodes" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">keep_unary=True</span></code> parameter meant that we kept intermediate (“unary”) nodes,
even those that do not not represent branch-points in the tree.
Often these are also unneeded, and by default we remove those too, although
this will mean that we lose track of the pedigree of the individuals
(which is stored in the parents column of the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-individual-table-definition" title="(in Project name not set)"><span>Individual Table</span></a>).
Since we are removing more nodes, the node IDs of non-samples will again change.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">current_day_genomes</span><span class="p">)</span>
<span class="n">simplified_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span> <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/217cf21a332517a3980da7266dc7c688075650e0a1da098f97f04d7affdb66a8.svg" src="_images/217cf21a332517a3980da7266dc7c688075650e0a1da098f97f04d7affdb66a8.svg" />
</div>
</div>
<p>This is now looking much more like a “normal” genetic genealogy (a “gene tree”),
in which all the sample genomes trace back to a single common ancestor.</p>
</section>
</section>
<section id="recombination">
<h2>Recombination<a class="headerlink" href="#recombination" title="Link to this heading">#</a></h2>
<p>If we pass a non-zero recombination rate to the <code class="docutils literal notranslate"><span class="pre">forward_WF()</span></code> function, different regions
of the genome may have different ancestries. This results in multiple trees along the genome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rho</span> <span class="o">=</span> <span class="mf">1e-7</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">forward_WF</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rho</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;A recombination rate of </span><span class="si">{</span><span class="n">rho</span><span class="si">}</span><span class="s2"> has created </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span><span class="si">}</span><span class="s2"> trees over </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>A recombination rate of 1e-07 has created 4 trees over 50000.0 bp
</pre></div>
</div>
</div>
</div>
<p>Here’s how the full (unsimplified) genealogy looks (labels omitted for clarity):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">graphics_params</span><span class="p">[</span><span class="s2">&quot;y_ticks&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">40</span> <span class="p">,</span><span class="mi">50</span><span class="p">]</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">400</span><span class="p">),</span> <span class="n">node_labels</span><span class="o">=</span><span class="p">{},</span> <span class="n">symbol_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="o">**</span><span class="n">graphics_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ec3f5029ec954dd343d391f31159b878ae98463750bbf5c4235d14d975bdfce8.svg" src="_images/ec3f5029ec954dd343d391f31159b878ae98463750bbf5c4235d14d975bdfce8.svg" />
</div>
</div>
<p>Because we are showing the extinct lineages and the recombinations associated with them,
this plot has many trees and nodes, and is rather confusing.
However, <a class="reference internal" href="#sec-tskit-forward-simulations-simplification"><span class="std std-ref">Simplification</span></a> can be equally applied to a
recombinant genealogy, and will reduce the genealogy to something more managable for
analysis and visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">forward_WF</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">generations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rho</span><span class="p">)</span>
<span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="n">graphics_params</span><span class="p">[</span><span class="s2">&quot;y_ticks&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">40</span> <span class="p">,</span><span class="mi">50</span><span class="p">]</span>
<span class="n">simplified_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span> <span class="o">**</span><span class="n">graphics_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/11349f1c08ed93c844e0556d13a2c6e4ff3e27dfe1323d558ec7fc6e25cb8fab.svg" src="_images/11349f1c08ed93c844e0556d13a2c6e4ff3e27dfe1323d558ec7fc6e25cb8fab.svg" />
</div>
</div>
<section id="larger-simulations">
<h3>Larger simulations<a class="headerlink" href="#larger-simulations" title="Link to this heading">#</a></h3>
<p>So far we have only simulated a relatively small population size (12 genomes / 6 diploids)
for a short time. We can easily simulate for longer times, and increase the population size.
In this case, simplification
can have a dramatic effect on the disk storage and memory required:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>

<span class="n">population_size</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">gens</span> <span class="o">=</span> <span class="mi">1000</span>  <span class="c1"># ten times the population size</span>
<span class="n">L</span> <span class="o">=</span> <span class="mi">500_000</span>  <span class="c1"># ten times the sequence length</span>

<span class="n">start</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
<span class="n">large_ts</span> <span class="o">=</span> <span class="n">forward_WF</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">gens</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rho</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Simulated </span><span class="si">{</span><span class="n">population_size</span><span class="si">}</span><span class="s2"> individuals (</span><span class="si">{</span><span class="n">population_size</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">2</span><span class="si">}</span><span class="s2"> genomes)&quot;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s2">&quot;for </span><span class="si">{</span><span class="n">gens</span><span class="si">}</span><span class="s2"> generations in </span><span class="si">{</span><span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="o">.</span><span class="n">seconds</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Simulated 100 individuals (200 genomes) for 1000 generations in 7.0 seconds
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Full tree sequence including dead lineages: </span><span class="si">{</span><span class="n">large_ts</span><span class="o">.</span><span class="n">nbytes</span><span class="o">/</span><span class="mi">1024</span><span class="o">/</span><span class="mi">1024</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB&quot;</span><span class="p">)</span>
<span class="n">current_day_genomes</span> <span class="o">=</span> <span class="n">large_ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">large_ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">current_day_genomes</span><span class="p">,</span> <span class="n">keep_input_roots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Tree sequence of current-day individuals: </span><span class="si">{</span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">nbytes</span><span class="o">/</span><span class="mi">1024</span><span class="o">/</span><span class="mi">1024</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB,&quot;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">num_trees</span><span class="si">}</span><span class="s2"> trees.&quot;</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Simplification has reduced the size by a factor of&quot;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">large_ts</span><span class="o">.</span><span class="n">nbytes</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">nbytes</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Full tree sequence including dead lineages: 16.79 MB
Tree sequence of current-day individuals: 0.05 MB, 113 trees.
Simplification has reduced the size by a factor of 320.57
</pre></div>
</div>
</div>
</div>
<p>The most obvious improvement when simulating genealogies in forward time
is therefore to carry out regular simplification steps <em>during</em> the simulation,
rather than just at the end. This is
described in the next tutorial:</p>
</section>
</section>
<section id="ensuring-coalescence">
<span id="sec-forward-simulations-ensuring-coalescence"></span><h2>Ensuring coalescence<a class="headerlink" href="#ensuring-coalescence" title="Link to this heading">#</a></h2>
<p>Even though the simulation above was run for hundreds of generations, there are still
trees with multiple roots in some regions of the genome. 1000 generations is not therefore
long enough to capture the ancestry back to a single common ancestor
(i.e. to ensure “full coalescence” of all local trees):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stairs</span><span class="p">(</span>
    <span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">num_roots</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">trees</span><span class="p">()],</span>
    <span class="n">simplified_ts</span><span class="o">.</span><span class="n">breakpoints</span><span class="p">(</span><span class="n">as_array</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">baseline</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Genome position (bp)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of roots&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/d539de123b1a91d80497f6c4b1dc5c6325523ebe624df08d6fb795d59be12fcc.png" src="_images/d539de123b1a91d80497f6c4b1dc5c6325523ebe624df08d6fb795d59be12fcc.png" />
</div>
</div>
<p>Most of our trees have one root, but a handful have two.
What is happening is that even though we simulated for <span class="math notranslate nohighlight">\(10N\)</span> generations,
and the theoretical expectation for the time to the most recent common ancestor
is <span class="math notranslate nohighlight">\(E[tMRCA]=4N\)</span>, the variance is pretty big, such that some marginal
trees are not completely coalesced.  Let’s take a look at the distribution of
tMRCAs using <code class="docutils literal notranslate"><span class="pre">msprime</span></code> to carry out an equivalent backward-time simulation
(which is much faster, so we can replicate it, say, 1000 times).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">tmrca</span><span class="o">=</span><span class="p">[]</span>
<span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">50_000</span>
<span class="n">rho</span> <span class="o">=</span> <span class="mf">5e-7</span>
<span class="k">for</span> <span class="n">sim</span> <span class="ow">in</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">100</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rho</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
    <span class="n">num_replicates</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
<span class="p">):</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">sim</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
        <span class="n">tmrca</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">tmrca</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;tMRCA (units of N generations)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Proportion of trees&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f99d6b35a1fd6bd05704ad32f3484100927e9b0e1f7c238510bfc7bf4242f043.png" src="_images/f99d6b35a1fd6bd05704ad32f3484100927e9b0e1f7c238510bfc7bf4242f043.png" />
</div>
</div>
<p>So, some fraction of local trees have very large tMRCAs! And of course
the larger the population, the longer the time needed to ensure full coalescence.
Hence for large forward simulation models, the number of generations required
to ensure full coalescence can be prohibitive.</p>
<p>Uncoalesced local trees matter, because they result in some pairs of
samples which are not connected, and so the genetic distance betweem
them is undefined (another way to imagine this is that we cannot spot
mutations that occur above the local roots of the tree). Hence the
simulation is not capturing the full genetic diversity within the sample.</p>
<p>A powerful way to get around this problem is <em>recapitation</em>,
in which an alternative technique, such as backward-in-time coalescent simulation,
is used to to fill in the “head” of the tree sequence. In other words,
we can use a fast backward-time simulator such as <code class="docutils literal notranslate"><span class="pre">msprime</span></code> to simulate the
prior genealogy of the oldest nodes in the simplified tree sequence.
Details are described in the <a class="reference internal" href="completing_forward_sims.html#sec-completing-forwards-simulations"><span class="std std-ref">Completing forwards simulations</span></a>
tutorial.</p>
</section>
<section id="more-complex-forward-simulations">
<h2>More complex forward-simulations<a class="headerlink" href="#more-complex-forward-simulations" title="Link to this heading">#</a></h2>
<p>The next tutorial shows the principles behind more complex simulations,
including e.g. regular simplification during the simulation,
adding mutations, and adding metadata.
It also details several extra tips and tricks we have learned when building forward simulators.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="completing_forward_sims.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Completing forwards simulations</p>
      </div>
    </a>
    <a class="right-next"
       href="more_forward_sims.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><em>Advanced forward simulations</em></p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#definitions">Definitions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#approach">Approach</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#core-steps">Core steps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inheritance-without-recombination">Inheritance without recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inheritance-with-recombination">Inheritance with recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-examples">Basic examples</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simplification">Simplification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-tskit-forward-simulations-simplification-id-changes">ID changes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-intermediate-nodes">Removing intermediate nodes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination">Recombination</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#larger-simulations">Larger simulations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ensuring-coalescence">Ensuring coalescence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#more-complex-forward-simulations">More complex forward-simulations</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>