
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Tables and editing &#8212; Tree Sequence Tutorials</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/tskit-book-theme.css?digest=4e121f05f62280197eeb3e8760a1903920f0778f" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/tskit-book-theme.js?digest=9a6366feaade5ec470560844be9ff19788eb9b85"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Simplification" href="simplification.html" />
    <link rel="prev" title="Parallelization" href="parallelization.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/tskit_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Tree Sequence Tutorials</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome!
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="what_is.html">
   What is a tree sequence?
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tree sequence basics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="terminology_and_concepts.html">
   Terminology &amp; concepts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="getting_started.html">
   Getting started with
   <strong class="program">
    tskit
   </strong>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistics and Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="analysing_tree_sequences.html">
   <em>
    Analysing tree sequences
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysing_trees.html">
   Analysing trees
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="counting_topologies.html">
   <em>
    Counting topologies
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="parallelization.html">
   <em>
    Parallelization
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Further tskit tutorials
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Tables and editing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="simplification.html">
   <em>
    Simplification
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="viz.html">
   Visualization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   Working with Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="args.html">
   <em>
    ARGs as tree sequences
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Simulation
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="simulation_overview.html">
   Tree sequences and simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="no_mutations.html">
   Do you really need mutations?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="completing_forward_sims.html">
   Completing forwards simulations
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="advanced_msprime.html">
   Advanced
   <strong class="program">
    msprime
   </strong>
   topics
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="demography.html">
     Demography
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="bottlenecks.html">
     Instantaneous Bottlenecks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="introgression.html">
     Introgression
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="forward_sims.html">
   <em>
    Building a forward simulator
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Other languages
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tskitr.html">
   Tskit and R
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Development
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial_development.html">
   <em>
    Developing new tutorials
   </em>
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/tskit-dev/tutorials"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/tskit-dev/tutorials/issues/new?title=Issue%20on%20page%20%2Ftables_and_editing.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/tables_and_editing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download notebook file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        <a href="_sources/tables_and_editing.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#correspondence-between-tables-and-trees">
   Correspondence between tables and trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#node-and-edge-tables">
     Node and edge tables
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#site-and-mutation-tables">
     Site and mutation tables
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tables-overview">
   Tables overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#individual-and-population-tables">
     Individual and population tables
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#metadata">
     Metadata
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#accessing-table-data">
   Accessing table data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#column-access">
     Column access
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#row-access">
     Row access
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#turning-tables-into-a-tree-sequence">
   Turning tables into a tree sequence
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sorting">
     Sorting
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#constructing-a-tree-sequence">
   Constructing a tree sequence
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#editing-tree-sequences">
   Editing tree sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#minor-edits">
     Minor edits
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#major-edits">
     Major edits
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Tables and editing</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#correspondence-between-tables-and-trees">
   Correspondence between tables and trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#node-and-edge-tables">
     Node and edge tables
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#site-and-mutation-tables">
     Site and mutation tables
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tables-overview">
   Tables overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#individual-and-population-tables">
     Individual and population tables
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#metadata">
     Metadata
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#accessing-table-data">
   Accessing table data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#column-access">
     Column access
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#row-access">
     Row access
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#turning-tables-into-a-tree-sequence">
   Turning tables into a tree sequence
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sorting">
     Sorting
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#constructing-a-tree-sequence">
   Constructing a tree sequence
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#editing-tree-sequences">
   Editing tree sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#minor-edits">
     Minor edits
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#major-edits">
     Major edits
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="tables-and-editing">
<span id="sec-tables"></span><h1>Tables and editing<a class="headerlink" href="#tables-and-editing" title="Permalink to this headline">#</a></h1>
<p>Internally, a tree sequence can be thought of as a set of tables, and <strong class="program">tskit</strong>
provides an interface for dealing with these tables directly. This is particularly
relevant when you wish to <a class="reference internal" href="#sec-tables-editing"><span class="std std-ref">edit or otherwise modify</span></a>
a tree sequence, although tables are also useful for bulk access to data contained in
the tree sequence, such as the times of all nodes.</p>
<p>There are eight tables that together define a tree sequence, although some may be empty,
and together they form a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.
The tables are defined in the official <strong class="program">tskit</strong> documentation for
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-table-definitions" title="(in Python)"><span class="xref std std-ref">Table Definitions</span></a>, and the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#sec-tables-api" title="(in Python)"><span class="xref std std-ref">Tables API</span></a> section in the docs describes how to work with
them. In this tutorial we give some pointers about what you can and cannot do with them.</p>
<section id="correspondence-between-tables-and-trees">
<span id="sec-tables-correspondence"></span><h2>Correspondence between tables and trees<a class="headerlink" href="#correspondence-between-tables-and-trees" title="Permalink to this headline">#</a></h2>
<p>Consider the following sequence of trees:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;data/tables_example.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tables_and_editing_4_0.svg" src="_images/tables_and_editing_4_0.svg" /></div>
</div>
<p>Ancestral recombination events have produced three different trees
that relate the three sampled genomes <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, and <code class="docutils literal notranslate"><span class="pre">2</span></code> to each other
along the chromosome of length 100.</p>
<section id="node-and-edge-tables">
<span id="sec-tables-correspondence-nodes-and-edges"></span><h3>Node and edge tables<a class="headerlink" href="#node-and-edge-tables" title="Permalink to this headline">#</a></h3>
<p>Each node in each of the above trees represents a particular ancestral genome
(a <em>haploid</em> genome; diploid individuals would be represented by two nodes).
Details about each node, including the time it lived, are stored in a
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NodeTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> (details <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-mutation-table-definition" title="(in Python)"><span class="xref std std-ref">here</span></a>)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>3</td><td>0</td><td>0</td><td>2</td><td>0.15000000</td><td></td></tr>
<tr><td>4</td><td>0</td><td>-1</td><td>-1</td><td>0.60000000</td><td></td></tr>
<tr><td>5</td><td>0</td><td>-1</td><td>-1</td><td>0.80000000</td><td></td></tr>
<tr><td>6</td><td>0</td><td>-1</td><td>-1</td><td>1.00000000</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>Importantly, the first column, <code class="docutils literal notranslate"><span class="pre">id</span></code>, is not actually recorded, and is
only shown when printing out node tables (as here) for convenience.
The second column, <code class="docutils literal notranslate"><span class="pre">flags</span></code> records a <code class="docutils literal notranslate"><span class="pre">1</span></code> for the nodes that are <em>samples</em>,
i.e., whose entire genealogical history is recorded by these trees.
You can see that the sample nodes are present in all the trees, but other nodes need
not be: for example node 3 is only seen in the middle tree. The <code class="docutils literal notranslate"><span class="pre">individual</span></code> and
<code class="docutils literal notranslate"><span class="pre">population</span></code> columns store an index into the rows of the
<a class="reference internal" href="#sec-tables-overview-individuals-and-populations"><span class="std std-ref">Individual and population tables</span></a>, or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> (<code class="docutils literal notranslate"><span class="pre">-1</span></code>)
if the node is not associated with an individual or population. The <code class="docutils literal notranslate"><span class="pre">time</span></code> column
records the historical age of the node, with contemporary nodes having a <code class="docutils literal notranslate"><span class="pre">time</span></code> of 0.</p>
<p>The way the nodes are related to each other (i.e. the tree topology) is stored
in an <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.EdgeTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> (details <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Python)"><span class="xref std std-ref">here</span></a>).
Since some edges are present in more than one tree (e.g., node 1 inherits from node 4
across the entire sequence), each edge contains not only the IDs of a parent and a child
node, but also the left and right positions which define the genomic region for which it
appears in the trees:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">edges</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>left</th><th>right</th><th>parent</th><th>child</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>20</td><td>80</td><td>3</td><td>0</td><td></td></tr>
<tr><td>1</td><td>20</td><td>80</td><td>3</td><td>2</td><td></td></tr>
<tr><td>2</td><td>0</td><td>100</td><td>4</td><td>1</td><td></td></tr>
<tr><td>3</td><td>0</td><td>20</td><td>4</td><td>2</td><td></td></tr>
<tr><td>4</td><td>80</td><td>100</td><td>4</td><td>2</td><td></td></tr>
<tr><td>5</td><td>20</td><td>80</td><td>4</td><td>3</td><td></td></tr>
<tr><td>6</td><td>80</td><td>100</td><td>5</td><td>0</td><td></td></tr>
<tr><td>7</td><td>80</td><td>100</td><td>5</td><td>4</td><td></td></tr>
<tr><td>8</td><td>0</td><td>20</td><td>6</td><td>0</td><td></td></tr>
<tr><td>9</td><td>0</td><td>20</td><td>6</td><td>4</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>Since node 3 is most recent, the edge that says that nodes 0 and 2 inherit
from node 3 on the interval between 0.2 and 0.8 comes first.  Next are the
edges from node 4: there are four of these, as the edge from node 4 to node
1 is shared across the entire sequence, and for each of the three
genomic intervals there is an additional child node. At this
point, we know the full tree on the middle interval.  Finally, edges
specifying the common ancestor of 0 and 4 on the remaining intervals (parents 6
and 5 respectively) allow us to construct all trees across the entire interval.</p>
</section>
<section id="site-and-mutation-tables">
<span id="sec-tables-correspondence-sites-and-mutations"></span><h3>Site and mutation tables<a class="headerlink" href="#site-and-mutation-tables" title="Permalink to this headline">#</a></h3>
<p>Most tree sequences have DNA variation data associated with them,
<a class="reference internal" href="what_is.html#sec-what-is-dna-data"><span class="std std-ref">stored as mutations overlaid on the trees</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;data/tables_example_muts.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mut_labels</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># An array of labels for the mutations, listing position &amp; allele change</span>
<span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>  <span class="c1"># This entire loop is just to make pretty labels</span>
    <span class="n">site</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">site</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">site</span><span class="p">)</span>
    <span class="n">older_mut</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">parent</span> <span class="o">&gt;=</span> <span class="mi">0</span>  <span class="c1"># is there an older mutation at the same position?</span>
    <span class="n">prev</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutation</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">parent</span><span class="p">)</span><span class="o">.</span><span class="n">derived_state</span> <span class="k">if</span> <span class="n">older_mut</span> <span class="k">else</span> <span class="n">site</span><span class="o">.</span><span class="n">ancestral_state</span>
    <span class="n">mut_labels</span><span class="p">[</span><span class="n">mut</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">→</span><span class="si">{}</span><span class="s2"> @</span><span class="si">{:g}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">prev</span><span class="p">,</span> <span class="n">mut</span><span class="o">.</span><span class="n">derived_state</span><span class="p">,</span> <span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mutation_labels</span><span class="o">=</span><span class="n">mut_labels</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tables_and_editing_11_0.svg" src="_images/tables_and_editing_11_0.svg" /></div>
</div>
<p>There are four mutations in the depiction above,
marked by red crosses: one above node <code class="docutils literal notranslate"><span class="pre">4</span></code> on the first tree which records an A to G
transition at position 15, another above node <code class="docutils literal notranslate"><span class="pre">1</span></code> in the second tree which records a G
to A transition at position 42, and the final two above nodes <code class="docutils literal notranslate"><span class="pre">0</span></code> and <code class="docutils literal notranslate"><span class="pre">3</span></code> recording
transitions, both at position 60, on the second tree. The positions are recorded in the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.SiteTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> (details <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-site-table-definition" title="(in Python)"><span class="xref std std-ref">here</span></a>):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>position</th><th>ancestral_state</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>15</td><td>A</td><td></td></tr>
<tr><td>1</td><td>42</td><td>G</td><td></td></tr>
<tr><td>2</td><td>60</td><td>T</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>As with node tables, the <code class="docutils literal notranslate"><span class="pre">id</span></code> column is <strong>not</strong> actually recorded, but is
implied by the position in the table.  The mutations themselves are recorded in the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.MutationTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> (details <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-mutation-table-definition" title="(in Python)"><span class="xref std std-ref">here</span></a>).
This associates each mutation with a site ID, the ID of the node above which the
mutation occurs, the derived state to which the allele has mutated, and (optionally) a
time at which the mutation occured.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>site</th><th>node</th><th>time</th><th>derived_state</th><th>parent</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td>4</td><td>0.90000000</td><td>G</td><td>-1</td><td></td></tr>
<tr><td>1</td><td>1</td><td>1</td><td>0.40000000</td><td>A</td><td>-1</td><td></td></tr>
<tr><td>2</td><td>2</td><td>3</td><td>0.55000000</td><td>C</td><td>-1</td><td></td></tr>
<tr><td>3</td><td>2</td><td>0</td><td>0.10000000</td><td>T</td><td>2</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>Where there are multiple mutations at the same site, there can be mutations
stacked on top of each other. The “parent” column therefore contains the ID of the
mutation immediately above the current one at the same site, or -1 if there is no parent.</p>
<p>These sites and mutations allow us to calculate the DNA sequence, or haplotype, for each
of the sample nodes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">h</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">haplotypes</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sample </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">h</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample 0: AGT
Sample 1: GAT
Sample 2: GGC
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="tables-overview">
<span id="sec-tables-overview"></span><h2>Tables overview<a class="headerlink" href="#tables-overview" title="Permalink to this headline">#</a></h2>
<p>Now that we’ve introduced the most important tables that make up a tree sequence, it’s
worth seeing how they come together to form a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For simplicity, this diagram omits 2 extra tables which form part of a
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>: the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.MigrationTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>
(recording details of migrations between populations), and the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.ProvenanceTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a>
(recording the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/provenance.html#sec-provenance" title="(in Python)"><span class="xref std std-ref">provenance</span></a> of the data in a tree sequence)</p>
</div>
</aside>
<p>The following visualization shows how they relate to one another, as well as introducing
two additional tables: the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> and <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.PopulationTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>
(which record which
<a class="reference internal" href="terminology_and_concepts.html#sec-terminology-individuals-and-populations"><span class="std std-ref">individuals and populations</span></a> a node
is contained in).</p>
<p><img alt="The tables in a tree sequence, and their interdependencies" src="_images/tables_uml.svg" /></p>
<p>The lines in this diagram below show where one table refers to the IDs
of another table, that is, they indicate dependencies between tables.</p>
<section id="individual-and-population-tables">
<span id="sec-tables-overview-individuals-and-populations"></span><h3>Individual and population tables<a class="headerlink" href="#individual-and-population-tables" title="Permalink to this headline">#</a></h3>
<p>Previously we noted that a row in the node table could contain an index into the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> and the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.PopulationTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>. In fact, in our example
tree sequence, the <a class="reference internal" href="#sec-tables-correspondence-nodes-and-edges"><span class="std std-ref">node table above</span></a>
indicates that nodes 0 to 3 belong to the population with ID 0 in the population table.
Here’s what that table looks like in this tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">populations</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div>
    <style scoped="">
        .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
        .tskit-table tbody tr th {vertical-align: top;}
        .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
        .tskit-table tbody th {padding: 0.5em 0.5em;}
    </style>
    <table border="1" class="tskit-table">
        <thead>
            <tr>
                <th>id</th><th>metadata</th>
            </tr>
        </thead>
        <tbody>
            <tr><td>0</td><td>{'name': 'GBR', 'description': 'British in England and Scotland'}</td></tr>

        </tbody>
    </table>
</div>
</div></div>
</div>
<p>The data in this table, and indeed in the individual table, is not visible in the
default tree sequence visualization, but can be very useful for keeping track of
information about the identity and location of nodes in the tree sequence. In
particular, you can see that the populations in this tree sequence have associated
<em>metadata</em>, a topic we will discuss in the
<a class="reference internal" href="#sec-tables-overview-metadata"><span class="std std-ref">next section</span></a>.</p>
<p>The node table also indicated that two nodes (node 0 and 1) point to the individual with
ID 0, one node (node 2) points to individual 1, and another node points to individual 2.
Here’s the relevant individual table:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>location</th><th>parents</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Alice'}</td></tr>
<tr><td>1</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Bob'}</td></tr>
<tr><td>2</td><td>0</td><td></td><td>-1, -1</td><td>{'name': 'Carol'}</td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>Since individual 0 (Alice) is referenced by two nodes (genomes), and assuming that this
is a diploid species, this indicates that we have sampled both maternal and paternal
genomes for this individual. The other two indivividuals, Bob and Carol, are referenced
by only one node, so either we only managed to sample one of their two genomes,
or alternatively they could represent haploid individuals.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">location</span></code> column of the individual table is intended for geographical location,
if relevant. The <code class="docutils literal notranslate"><span class="pre">parents</span></code> column, if filled out, allows us to construct
parts of the <em>pedigree</em> (the genealogical family tree) connecting the individuals
in question. It contains a list of individual IDs (or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code>/<code class="docutils literal notranslate"><span class="pre">-1</span></code> if a
parent is unknown). In this case, the <code class="docutils literal notranslate"><span class="pre">parents</span></code> column indicates that Alice and Bob
share the same parent, individual 2 (i.e. Carol). That means Alice and Bob must be
either siblings or half siblings.</p>
</section>
<section id="metadata">
<span id="sec-tables-overview-metadata"></span><h3>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">#</a></h3>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more information on schemas, see the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata" title="(in Python)"><span>Metadata</span></a> section of the
official <strong class="program">tskit</strong> documentation</p>
</div>
</aside>
<p>As can be seen above, each table row, as well as the entire table collection, can be
associated with (optional) metadata, which can be stored in various formats. The format
is determined by the <em>schema</em> of the metadata. In this example, for clarity, the
metadata of the population and individual tables is stored as
<a class="reference external" href="https://www.json.org/">JSON</a>.</p>
<p>Because we have chosen to use JSON for the metadata of the individuals table, it is
easy to access the metadata values:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Information from the tree sequence&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">individuals</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Individual </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> is named </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Equivalent information from the associated tables&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Individual name: </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Information from the tree sequence
Individual 0 is named Alice
Individual 1 is named Bob
Individual 2 is named Carol

Equivalent information from the associated tables
Individual name: Alice
Individual name: Bob
Individual name: Carol
</pre></div>
</div>
</div>
</div>
<p><a class="reference internal" href="#sec-tables-editing-minor"><span class="std std-ref">Later in this tutorial</span></a> there are some basic
examples of editing the metadata fields; there is also an extensive
<a class="reference internal" href="metadata.html#sec-tutorial-metadata"><span class="std std-ref">Working with Metadata</span></a> tutorial which provides much more
detail about how to construct and use <code class="docutils literal notranslate"><span class="pre">tskit</span></code> metadata.</p>
</section>
</section>
<section id="accessing-table-data">
<span id="sec-tables-accessing"></span><h2>Accessing table data<a class="headerlink" href="#accessing-table-data" title="Permalink to this headline">#</a></h2>
<p>A simple way to see the contents of a table is to <code class="docutils literal notranslate"><span class="pre">print</span></code> it:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔══╤═════╤══════════╤══════════╤══════════╤════════╗
║id│flags│population│individual│time      │metadata║
╠══╪═════╪══════════╪══════════╪══════════╪════════╣
║0 │    1│         0│         0│0.00000000│        ║
║1 │    1│         0│         0│0.00000000│        ║
║2 │    1│         0│         1│0.00000000│        ║
║3 │    0│         0│         2│0.15000000│        ║
║4 │    0│        -1│        -1│0.60000000│        ║
║5 │    0│        -1│        -1│0.80000000│        ║
║6 │    0│        -1│        -1│1.00000000│        ║
╚══╧═════╧══════════╧══════════╧══════════╧════════╝
</pre></div>
</div>
</div>
</div>
<p>But <strong class="program">tskit</strong> also provides access to columns and rows separately.</p>
<section id="column-access">
<span id="sec-tables-accessing-columns"></span><h3>Column access<a class="headerlink" href="#column-access" title="Permalink to this headline">#</a></h3>
<p>An entire column in a table can be extracted as a <strong class="program">numpy</strong> array from the table
object. For instance, if <code class="docutils literal notranslate"><span class="pre">n</span></code> is a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NodeTable" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>, then <code class="docutils literal notranslate"><span class="pre">n.time</span></code>
will return an array containing the birth times of the individuals whose genomes
are represented by the nodes in the table.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0.  , 0.  , 0.  , 0.15, 0.6 , 0.8 , 1.  ])
</pre></div>
</div>
</div>
</div>
<p>You can also <em>modify</em> entire columns, but when doing this to tables from a tree
sequence, you should always modify a <em>copy</em> of the original data, using the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.dump_tables" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump_tables()</span></code></a> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>  <span class="c1"># make a copy of the tree sequence tables, for editing</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0.  , 0.  , 0.  , 0.15, 0.6 , 0.8 , 1.  ])
</pre></div>
</div>
</div>
</div>
<p>Moreover, it is important to note that this numpy array itself is copy of the data
stored in <code class="docutils literal notranslate"><span class="pre">new_tables</span></code>. In other words, modifying individual elements of <code class="docutils literal notranslate"><span class="pre">n.time</span></code> will
<em>not</em> change the node table <code class="docutils literal notranslate"><span class="pre">n</span></code>. To change the column data, you need to take a copy
of the array, modify it, and assign it back in. For example, here we add 0.25 to
every <code class="docutils literal notranslate"><span class="pre">time</span></code> except the first in the node table:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">node_table</span> <span class="o">=</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">nodes</span>
<span class="n">times</span> <span class="o">=</span> <span class="n">node_table</span><span class="o">.</span><span class="n">time</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Old node times:&quot;</span><span class="p">,</span> <span class="n">times</span><span class="p">)</span>
<span class="n">times</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="n">times</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="mf">0.25</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">times</span>
<span class="n">node_table</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Old node times: [0.   0.   0.   0.15 0.6  0.8  1.  ]
</pre></div>
</div>
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.25000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.25000000</td><td></td></tr>
<tr><td>3</td><td>0</td><td>0</td><td>2</td><td>0.40000000</td><td></td></tr>
<tr><td>4</td><td>0</td><td>-1</td><td>-1</td><td>0.85000000</td><td></td></tr>
<tr><td>5</td><td>0</td><td>-1</td><td>-1</td><td>1.05000000</td><td></td></tr>
<tr><td>6</td><td>0</td><td>-1</td><td>-1</td><td>1.25000000</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>When assigning columns like this, an error will be raised if the column is not of the
expected length:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">node_table</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">times</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span> <span class="p">[</span><span class="mi">18</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">node_table</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">times</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>

<span class="nn">File /opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/tskit/tables.py:489,</span> in <span class="ni">BaseTable.__setattr__</span><span class="nt">(self, name, value)</span>
<span class="g g-Whitespace">    </span><span class="mi">487</span>     <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">asdict</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">488</span>     <span class="n">d</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
<span class="ne">--&gt; </span><span class="mi">489</span>     <span class="bp">self</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="o">**</span><span class="n">d</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">490</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">491</span>     <span class="nb">object</span><span class="o">.</span><span class="fm">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

<span class="nn">File /opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/tskit/tables.py:1202,</span> in <span class="ni">NodeTable.set_columns</span><span class="nt">(self, flags, time, population, individual, metadata, metadata_offset, metadata_schema)</span>
<span class="g g-Whitespace">   </span><span class="mi">1169</span> <span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1170</span><span class="sd"> Sets the values for each column in this :class:`NodeTable` using the values in</span>
<span class="g g-Whitespace">   </span><span class="mi">1171</span><span class="sd"> the specified arrays. Overwrites existing data in all the table columns.</span>
<span class="sd">   (...)</span>
<span class="g g-Whitespace">   </span><span class="mi">1199</span><span class="sd"> :type metadata_schema: str</span>
<span class="g g-Whitespace">   </span><span class="mi">1200</span><span class="sd"> &quot;&quot;&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1201</span> <span class="bp">self</span><span class="o">.</span><span class="n">_check_required_args</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">flags</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="n">time</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1202</span> <span class="bp">self</span><span class="o">.</span><span class="n">ll_table</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1203</span>     <span class="nb">dict</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1204</span>         <span class="n">flags</span><span class="o">=</span><span class="n">flags</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1205</span>         <span class="n">time</span><span class="o">=</span><span class="n">time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1206</span>         <span class="n">population</span><span class="o">=</span><span class="n">population</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1207</span>         <span class="n">individual</span><span class="o">=</span><span class="n">individual</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1208</span>         <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1209</span>         <span class="n">metadata_offset</span><span class="o">=</span><span class="n">metadata_offset</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1210</span>         <span class="n">metadata_schema</span><span class="o">=</span><span class="n">metadata_schema</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1211</span>     <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1212</span> <span class="p">)</span>

<span class="ne">ValueError</span>: Input array dimensions must be equal.
</pre></div>
</div>
</div>
</div>
</section>
<section id="row-access">
<span id="sec-tables-accessing-rows"></span><h3>Row access<a class="headerlink" href="#row-access" title="Permalink to this headline">#</a></h3>
<p>Rows can be accessed using square braces, which will return an object containing the
raw values, accessible via attributes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">row</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Node 3, belonging to individual&quot;</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">individual</span><span class="p">,</span>
    <span class="s2">&quot;in population&quot;</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">population</span><span class="p">,</span>
    <span class="s2">&quot;exists at time&quot;</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">time</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node 3, belonging to individual 2 in population 0 exists at time 0.15
NodeTableRow(flags=0, time=0.15, population=0, individual=2, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Additionally, many rows can be extracted into a new table using slices or
<a class="reference external" href="https://numpy.org/doc/stable/user/basics.indexing.html#basics-indexing" title="(in NumPy v1.23)"><span class="xref std std-ref">numpy indexing</span></a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>0</td><td>0</td><td>2</td><td>0.15000000</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>As with columns, to change the data in a row of a tree sequence table, you must operate
on the copy that was produced by <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.dump_tables" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump_tables()</span></code></a>. Here we’ll try to
change the ancestral state value of site 0.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Check the site&#39;s ancestral state (&quot;A&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SiteTableRow(position=15.0, ancestral_state=&#39;A&#39;, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Since the row object(s) returned by the calls above are themselves copies,
the attributes enforce read-only access:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ancestral_state</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span> <span class="c1"># This will not work: attributes are read-only</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FrozenInstanceError</span><span class="g g-Whitespace">                       </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span> <span class="p">[</span><span class="mi">22</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ancestral_state</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span>

<span class="nn">File &lt;string&gt;:4,</span> in <span class="ni">__setattr__</span><span class="nt">(self, name, value)</span>

<span class="ne">FrozenInstanceError</span>: cannot assign to field &#39;ancestral_state&#39;
</pre></div>
</div>
</div>
</div>
<p>So to change row data you need to replace the row completely:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTableRow</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>For convenience, each row object has a <code class="docutils literal notranslate"><span class="pre">replace</span></code> method which creates a new table row
from an old one. This is often an easier way to changing one or two attributes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Another way to change a single entry: create a new row to replace the old one ...</span>
<span class="n">replacement_row</span> <span class="o">=</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">ancestral_state</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
<span class="c1"># ... then overwrite the original</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">replacement_row</span>

<span class="c1"># Or equivalently, do both in a single step</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">ancestral_state</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>It is also easy to add more rows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_pos</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">new_site_id</span> <span class="o">=</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span>
    <span class="n">position</span><span class="o">=</span><span class="n">new_pos</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="sa">b</span><span class="s2">&quot;An empty site&quot;</span>
    <span class="c1"># NB: For this example table we have to feed a raw byte string as metadata</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New empty site allocated at position </span><span class="si">{</span><span class="n">new_pos</span><span class="si">}</span><span class="s2"> with ID </span><span class="si">{</span><span class="n">new_site_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">sites</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>New empty site allocated at position 10 with ID 3
</pre></div>
</div>
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>position</th><th>ancestral_state</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>15</td><td>C</td><td></td></tr>
<tr><td>1</td><td>42</td><td>G</td><td></td></tr>
<tr><td>2</td><td>60</td><td>T</td><td></td></tr>
<tr><td>3</td><td>10</td><td>G</td><td>b'An empty site'</td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
</section>
</section>
<section id="turning-tables-into-a-tree-sequence">
<span id="sec-tables-to-tree-sequence"></span><h2>Turning tables into a tree sequence<a class="headerlink" href="#turning-tables-into-a-tree-sequence" title="Permalink to this headline">#</a></h2>
<p>The <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection.tree_sequence" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> method will attempt to turn a table collection
into a tree sequence. This is not guaranteed to work: it’s possible you have created a
nonsensical tree sequence where, for example, a child node has multiple parents at
a given position, or where edges reference non-existant nodes. The
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-valid-tree-sequence-requirements" title="(in Python)"><span>Valid tree sequence requirements</span></a> section of the official tskit docs
lists the requirements for a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> to represent a valid
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p>
<section id="sorting">
<span id="sec-tables-to-tree-sequence-sorting"></span><h3>Sorting<a class="headerlink" href="#sorting" title="Permalink to this headline">#</a></h3>
<p>Even if the tables represent a valid tree sequence, for efficiency reasons tree
sequences also require several of their constituent tables to be sorted in a specific
order. For instance, edges are <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-requirements" title="(in Python)"><span class="xref std std-ref">required</span></a> to be sorted
in nondecreasing order of parent time, and sites are
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-site-requirements" title="(in Python)"><span class="xref std std-ref">required</span></a> to be sorted in order of position.
We can ensure that a set of tables are correctly sorted by calling the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection.sort" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> method.</p>
<p>In fact the new empty site that we added to the <code class="docutils literal notranslate"><span class="pre">new_tables</span></code> collection in the previous
section was at the start of the genome, so the table collection no longer has its sites
in the order required for a tree sequence. That means we need to call
<code class="docutils literal notranslate"><span class="pre">new_tables.sort()</span></code> before turning the table collection into a tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># altered_ts = new_tables.tree_sequence()  # This won&#39;t work</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">altered_ts</span> <span class="o">=</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>  <span class="c1"># Now it will</span>
</pre></div>
</div>
</div>
</div>
<p>Now that it had been turned into a tree sequence, we can plot it:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot without mutation labels, for clarity </span>
<span class="n">altered_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">y_gridlines</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mutation_labels</span><span class="o">=</span><span class="p">{})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tables_and_editing_51_0.svg" src="_images/tables_and_editing_51_0.svg" /></div>
</div>
<p>You can see that the new tree sequence has been modified as expected: there is a new
empty site at position 10 (represented by a tickmark above the X axis with no mutations
on it), and all the nodes except node 0 have had their times increased by 0.25.</p>
</section>
</section>
<section id="constructing-a-tree-sequence">
<span id="sec-tables-creating"></span><h2>Constructing a tree sequence<a class="headerlink" href="#constructing-a-tree-sequence" title="Permalink to this headline">#</a></h2>
<p>With the tools above in mind, we can now see how to construct a tree sequence by hand.
It’s unlikely that you would ever need to do this from scratch, but it’s helpful to
understand how tables work. We’ll build an extremely simple tree sequence, consisting of
a single tree that looks like this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;data/construction_example.trees&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tables_and_editing_53_0.svg" src="_images/tables_and_editing_53_0.svg" /></div>
</div>
<p>Starting with an empty set of tables, we can fill, say, the node information by using
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NodeTable.add_row" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">NodeTable.add_row()</span></code></a> as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="n">sequence_length</span><span class="o">=</span><span class="mf">1e3</span><span class="p">)</span>
<span class="n">node_table</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>  <span class="c1"># set up an alias, for efficiency</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span><span class="p">)</span>  <span class="c1"># Node 0: defaults to time 0</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span><span class="p">)</span>  <span class="c1"># Node 1: defaults to time 0</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Node 2: not a sample</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span><span class="p">)</span>  <span class="c1"># Node 3</span>
<span class="n">node_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>  <span class="c1"># Node 4</span>
<span class="n">node_table</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>1</td><td>-1</td><td>-1</td><td>0</td><td></td></tr>
<tr><td>1</td><td>1</td><td>-1</td><td>-1</td><td>0</td><td></td></tr>
<tr><td>2</td><td>0</td><td>-1</td><td>-1</td><td>10</td><td></td></tr>
<tr><td>3</td><td>1</td><td>-1</td><td>-1</td><td>0</td><td></td></tr>
<tr><td>4</td><td>0</td><td>-1</td><td>-1</td><td>20</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">.add_row()</span></code> method is natural (and should be reasonably efficient) if
new records appear one-by-one. Alternatively <code class="docutils literal notranslate"><span class="pre">.set_columns()</span></code> can be used to
set columns for all the rows at once, by passing in numpy arrays of the appropriate
type. We’ll use that for the edges:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">edge_table</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">edges</span>
<span class="n">edge_table</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span>
    <span class="n">left</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]),</span>
    <span class="n">right</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1e3</span><span class="p">,</span> <span class="mf">1e3</span><span class="p">,</span> <span class="mf">1e3</span><span class="p">,</span> <span class="mf">1e3</span><span class="p">]),</span>
    <span class="n">parent</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>  <span class="c1"># References IDs in the node table</span>
    <span class="n">child</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>  <span class="c1"># References IDs in the node table</span>
<span class="p">)</span>
<span class="n">edge_table</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>left</th><th>right</th><th>parent</th><th>child</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td>1000</td><td>2</td><td>0</td><td></td></tr>
<tr><td>1</td><td>0</td><td>1000</td><td>2</td><td>1</td><td></td></tr>
<tr><td>2</td><td>0</td><td>1000</td><td>4</td><td>2</td><td></td></tr>
<tr><td>3</td><td>0</td><td>1000</td><td>4</td><td>3</td><td></td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>And finally we can add a site and a mutation: here we’ll use 0/1 mutations rather than
ATGC.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">site_id</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="mf">500.0</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="s1">&#39;0&#39;</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">site_id</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">derived_state</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;A hand-built tree sequence!&quot;</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>A hand-built tree sequence!
</pre></div>
</div>
<img alt="_images/tables_and_editing_59_1.svg" src="_images/tables_and_editing_59_1.svg" /></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The ordering requirements for a valid tree sequence
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-node-requirements" title="(in Python)"><span class="xref std std-ref">do not specify</span></a> that rows in the
node table have to be sorted in any particular order, e.g. by time. By convention,
sample nodes are often the first ones listed in the node table, and this is the node
order returned by <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a>, but the example above shows that
sample nodes need not necessarily be those with the IDs <span class="math notranslate nohighlight">\(0..n\)</span>.</p>
</div>
</section>
<section id="editing-tree-sequences">
<span id="sec-tables-editing"></span><h2>Editing tree sequences<a class="headerlink" href="#editing-tree-sequences" title="Permalink to this headline">#</a></h2>
<p>If you may wish to make modifications to a tree sequence, you will immediately encounter
a barrier: tree sequence objects are <strong>immutable</strong> and so cannot be edited in-place.
Therefore, the way to change a tree sequence is to edit its underlying tables.</p>
<p>Before providing further examples of how to edit tables, it’s worth noting that
<strong class="program">tskit</strong> may provide a built-in method to achieve what you want to do, without
you having to think about the underlying tables at all. In particular, several methods
will return a new tree sequence that has been modified in some way. Here are some key
ones:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.delete_sites" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a> returns a tree sequence with certain sites
deleted</p></li>
<li><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.delete_intervals" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals()</span></code></a> and <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.keep_intervals" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals()</span></code></a>
return a tree sequence whose trees cover a smaller fraction of the genome (and which can
be combined with <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.trim" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trim()</span></code></a> to change the coordinate system)</p></li>
<li><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a> returns a new tree sequence with some sample nodes removed
(see the <a class="reference internal" href="simplification.html#sec-simplification"><span class="std std-ref">simplification tutorial</span></a>).</p></li>
<li><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.union" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.union()</span></code></a> returns a new tree sequence formed by merging two other tree
sequences together.</p></li>
</ul>
<p>For example, to strip singleton sites from a tree sequence, we can simply call the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.delete_sites" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="p">,</span> <span class="s2">&quot;sites in the simulated tree sequence&quot;</span><span class="p">)</span>

<span class="c1"># Identify sites with a singleton allele</span>
<span class="n">sites_with_a_singleton_allele</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="n">non_missing_genotypes</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">genotypes</span><span class="p">[</span><span class="n">v</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MISSING_DATA</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">non_missing_genotypes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">sites_with_a_singleton_allele</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        
<span class="c1"># Strip those sites from the tree sequence</span>
<span class="n">ts_new</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">delete_sites</span><span class="p">(</span><span class="n">sites_with_a_singleton_allele</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts_new</span><span class="o">.</span><span class="n">num_sites</span><span class="p">,</span> <span class="s2">&quot;sites after removing singletons&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>109 sites in the simulated tree sequence
52 sites after removing singletons
</pre></div>
</div>
</div>
</div>
<p>However, you want to do something not covered by a built-in method, you will
need to edit a copy of the underlying tables and then create a new tree sequence from
the modified tables.</p>
<section id="minor-edits">
<span id="sec-tables-editing-minor"></span><h3>Minor edits<a class="headerlink" href="#minor-edits" title="Permalink to this headline">#</a></h3>
<p><a class="reference internal" href="#sec-tables-accessing-rows"><span class="std std-ref">Previously</span></a> we saw how to change specific rows by
overwriting them. Here, for example, we use <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTableRow.replace" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IndividualTableRow.replace()</span></code></a> to change
the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> associated with an individual in one of our example tree sequences:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">altered_ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="c1"># Change the name &quot;Bob&quot; to &quot;Robert&quot;</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;Robert&quot;</span><span class="p">})</span>
<span class="n">edited_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">edited_ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>location</th><th>parents</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Alice'}</td></tr>
<tr><td>1</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Robert'}</td></tr>
<tr><td>2</td><td>0</td><td></td><td>-1, -1</td><td>{'name': 'Carol'}</td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
<p>If you are overwriting an entire column, especially something like metadata which is a
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-encoding-ragged-columns" title="(in Python)"><span class="xref std std-ref">ragged column</span></a>, looping through each row and changing
it as above will not be very performant. Instead, you should overwrite the entire column
(or multiple columns) at once. We <a class="reference internal" href="#sec-tables-accessing-columns"><span class="std std-ref">previously saw</span></a>
how to overwrite an entire column of numbers, and for most columns that should suffice.
However, overwriting ragged columns like <code class="docutils literal notranslate"><span class="pre">metadata</span></code> or <code class="docutils literal notranslate"><span class="pre">parents</span></code> is more involved,
as you cannot set them to a single numpy array. Instead, for these columns, <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
provides a number of <code class="docutils literal notranslate"><span class="pre">packset_</span></code> routines, which set the appropriate column from a Python
list of values.</p>
<p>Here’s an example of changing an entire column of metadata in one pass (in this case,
note the extra step of validating and encoding before calling
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTable.packset_metadata" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IndividualTable.packset_metadata()</span></code></a>, as
<a class="reference internal" href="metadata.html#sec-tutorial-metadata-bulk"><span class="std std-ref">described</span></a> in the
<a class="reference internal" href="metadata.html#sec-tutorial-metadata"><span class="std std-ref">metadata tutorial</span></a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add a surname to all individuals</span>
<span class="n">new_metadata</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; Smith&quot;</span><span class="p">}</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">]</span>
<span class="n">validated_metadata</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">validate_and_encode_row</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">new_metadata</span>
<span class="p">]</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">packset_metadata</span><span class="p">(</span><span class="n">validated_metadata</span><span class="p">)</span>
<span class="n">edited_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">edited_ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
                <style scoped="">
                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                    .tskit-table tbody tr th {vertical-align: top;}
                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                    .tskit-table tbody th {padding: 0.5em 0.5em;}
                </style>
                <table border="1" class="tskit-table">
                    <thead>
                        <tr>
                            <th>id</th><th>flags</th><th>location</th><th>parents</th><th>metadata</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr><td>0</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Alice Smith'}</td></tr>
<tr><td>1</td><td>0</td><td></td><td>2, -1</td><td>{'name': 'Robert Smith'}</td></tr>
<tr><td>2</td><td>0</td><td></td><td>-1, -1</td><td>{'name': 'Carol Smith'}</td></tr>

                    </tbody>
                </table>
            </div>
</div></div>
</div>
</section>
<section id="major-edits">
<span id="sec-tables-editing-major"></span><h3>Major edits<a class="headerlink" href="#major-edits" title="Permalink to this headline">#</a></h3>
<p>For larger changes, it is often worth clearing the entire table and filling it. Here,
for example, we clear all mutations from the mutations table, then fill it up again,
if possible using a more parsimonious placement of mutations (i.e. a placement that
would better explain the observed variation among samples - see the
<a class="reference internal" href="analysing_trees.html#sec-analysing-trees-parsimony"><span class="std std-ref">Parsimony</span></a> tutorial section).
This also involves changing the ancestral state at the site if necessary.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make a tree sequence with loads of recurrent/back mutations, for testing only</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">456</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Original ts has&quot;</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_mutations</span><span class="p">,</span> <span class="s2">&quot;mutations on&quot;</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span><span class="p">,</span> <span class="s2">&quot;trees&quot;</span><span class="p">)</span>

<span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>  <span class="c1"># Copy the table collection associated with this tree sequence</span>
<span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>  <span class="c1"># Clear all mutations in the table collection copy</span>
<span class="n">v_iter</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">()</span>  <span class="c1"># Workaround until https://github.com/tskit-dev/tskit/issues/605 is solved</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
        <span class="n">variant</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">v_iter</span><span class="p">)</span>
        <span class="n">anc_state</span><span class="p">,</span> <span class="n">mutations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span> <span class="o">&lt;</span>  <span class="nb">len</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">):</span>
            <span class="c1"># Genotypes can be explained with a more parsimonious distribution of mutations</span>
            <span class="k">if</span> <span class="n">anc_state</span> <span class="o">!=</span> <span class="n">site</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">:</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">ancestral_state</span><span class="o">=</span><span class="n">anc_state</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>            
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Add the original mutations back in (note that parent ids may have changed,</span>
            <span class="c1"># so we will need to call compute_mutation_parents at the end)</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">compute_mutation_parents</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">compute_mutation_times</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">parsimonious_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Parsimonious ts has&quot;</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_mutations</span> <span class="o">-</span> <span class="n">parsimonious_ts</span><span class="o">.</span><span class="n">num_mutations</span><span class="p">,</span> <span class="s2">&quot;fewer muts&quot;</span><span class="p">)</span>

<span class="c1"># Optional checks</span>
<span class="k">assert</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span> <span class="o">==</span> <span class="n">parsimonious_ts</span><span class="o">.</span><span class="n">num_trees</span>
<span class="k">assert</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span> <span class="o">==</span> <span class="n">parsimonious_ts</span><span class="o">.</span><span class="n">num_sites</span>
<span class="k">for</span> <span class="n">v1</span><span class="p">,</span> <span class="n">v2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">(),</span> <span class="n">parsimonious_ts</span><span class="o">.</span><span class="n">variants</span><span class="p">()):</span>
    <span class="c1"># convert alleles to numpy arrays so we can index into them easily</span>
    <span class="n">alleles1</span><span class="p">,</span> <span class="n">alleles2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">v1</span><span class="o">.</span><span class="n">alleles</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">v2</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
    <span class="c1"># Verify the distribution of variation is the same, by comparing the allelic state</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">alleles1</span><span class="p">[</span><span class="n">v1</span><span class="o">.</span><span class="n">genotypes</span><span class="p">]</span> <span class="o">==</span> <span class="n">alleles2</span><span class="p">[</span><span class="n">v2</span><span class="o">.</span><span class="n">genotypes</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Original ts has 154 mutations on 71 trees
Parsimonious ts has 33 fewer muts
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the code above we iterate over the trees in the original tree sequence, but make
changes to a <em>copy</em> of the tables associated with those trees, modifying the table
collection as we go. This is a common idiom.</p>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="parallelization.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><em>Parallelization</em></p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="simplification.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><em>Simplification</em></p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Tskit Developers<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>