
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Tskit and R &#8212; Tree Sequence Tutorials</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tskitr';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Developing new tutorials" href="tutorial_development.html" />
    <link rel="prev" title="Advanced forward simulations" href="more_forward_sims.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tskit_logo.svg" class="logo__image only-light" alt="Tree Sequence Tutorials - Home"/>
    <script>document.write(`<img src="_static/tskit_logo.svg" class="logo__image only-dark" alt="Tree Sequence Tutorials - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="what_is.html">What is a tree sequence?</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tree sequence basics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="terminology_and_concepts.html">Terminology &amp; concepts</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started with <strong class="program">tskit</strong></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistics and Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="analysing_tree_sequences.html"><em>Analysing tree sequences</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="analysing_trees.html">Analysing trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="incremental_algorithms.html"><em>Incremental algorithms</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="counting_topologies.html">Counting topologies</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallelization.html"><em>Parallelization</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Further tskit tutorials</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tables_and_editing.html">Tables and editing</a></li>
<li class="toctree-l1"><a class="reference internal" href="simplification.html"><em>Simplification</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Working with Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="args.html">ARGs as tree sequences</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="simulation_overview.html">Tree sequences and simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="no_mutations.html">Do you really need mutations?</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="advanced_msprime.html">Advanced <strong class="program">msprime</strong> topics</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="demography.html">Demography</a></li>
<li class="toctree-l2"><a class="reference internal" href="bottlenecks.html">Instantaneous Bottlenecks</a></li>
<li class="toctree-l2"><a class="reference internal" href="introgression.html">Introgression</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="completing_forward_sims.html">Completing forwards simulations</a></li>
<li class="toctree-l1"><a class="reference internal" href="forward_sims.html">Building a forward simulator</a></li>
<li class="toctree-l1"><a class="reference internal" href="more_forward_sims.html"><em>Advanced forward simulations</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Other languages</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Tskit and R</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tutorial_development.html"><em>Developing new tutorials</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tskit for ...</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="popgen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for population genetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for phylogenetics</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/tskitr.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Tskit and R</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#attributes-and-methods">Attributes and methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ids-and-indexes">IDs and indexes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis">Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#jupyter-notebook-tips">Jupyter notebook tips</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interaction-with-r-libraries">Interaction with R libraries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-information">Further information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-usage-through-the-tskit-s-c-api">Advanced usage through the tskit’s C API</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tskit-and-r">
<span id="sec-tskit-r"></span><h1>Tskit and R<a class="headerlink" href="#tskit-and-r" title="Link to this heading">#</a></h1>
<p>To interface with <code class="docutils literal notranslate"><span class="pre">tskit</span></code> in R, we can use the <a class="reference external" href="https://rstudio.github.io/reticulate">reticulate</a> R package, which lets you call <code class="docutils literal notranslate"><span class="pre">tskit</span></code>’s Python API within an R session. In this tutorial, we’ll go through a couple of examples to show you how to get started. You’ll need to install <code class="docutils literal notranslate"><span class="pre">reticulate</span></code> in your R session via <code class="docutils literal notranslate"><span class="pre">install.packages(&quot;reticulate&quot;)</span></code> and at a minimum state the required Python packages via <code class="docutils literal notranslate"><span class="pre">reticulate::py_require(c(&quot;tskit&quot;,</span> <span class="pre">&quot;msprime&quot;))</span></code>. This setting will use <code class="docutils literal notranslate"><span class="pre">reticulate</span></code>’s isolated Python virtual environment, but you can also use an alternative Python environment as described at https://rstudio.github.io/reticulate.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference external" href="https://www.slendr.net">slendr</a> R package uses <code class="docutils literal notranslate"><span class="pre">reticulate</span></code> extensively to work with tree sequences in R and also provides R wrappers for some <code class="docutils literal notranslate"><span class="pre">tskit</span></code> functions.</p>
</div>
</aside>
<p>We’ll begin by simulating a small tree sequence using <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">msprime</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reticulate</span><span class="o">::</span><span class="nf">import</span><span class="p">(</span><span class="s">&quot;msprime&quot;</span><span class="p">)</span>
<span class="n">ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">msprime</span><span class="o">$</span><span class="nf">sim_ancestry</span><span class="p">(</span><span class="m">80</span><span class="p">,</span><span class="w"> </span><span class="n">sequence_length</span><span class="o">=</span><span class="m">1e4</span><span class="p">,</span><span class="w"> </span><span class="n">recombination_rate</span><span class="o">=</span><span class="m">1e-4</span><span class="p">,</span><span class="w"> </span><span class="n">random_seed</span><span class="o">=</span><span class="m">42</span><span class="p">)</span>
<span class="n">ts</span><span class="w">  </span><span class="c1"># See &quot;Jupyter notebook tips&quot;, below for how to render this nicely</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;tskit.trees.TreeSequence object at 0x7f195a8d4190&gt;
</pre></div>
</div>
</div>
</div>
<section id="attributes-and-methods">
<h2>Attributes and methods<a class="headerlink" href="#attributes-and-methods" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">reticulate</span></code> allows us to access a Python object’s attributes via
the <code class="docutils literal notranslate"><span class="pre">$</span></code> operator. For example, we can access (and assign to a native R object)
the number of samples in the tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">num_samples</span>
<span class="n">n</span>
<span class="nf">is</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="w">  </span><span class="c1"># Show the type of the object</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">160</div><div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'integer'</li><li>'double'</li><li>'numeric'</li><li>'vector'</li><li>'data.frameRowLabels'</li><li>'index'</li><li>'replValue'</li><li>'numLike'</li><li>'number'</li><li>'atomicVector'</li><li>'replValueSp'</li></ol>
</div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">$</span></code> operator can also be used to call methods, for example, the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> method associated with the tree sequence.
The method parameters are given as native R objects
(but note that object IDs still use tskit’s 0-based indexing system).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">simplify</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">7</span><span class="p">)</span><span class="w">  </span><span class="c1"># only keep samples with ids 0, 1, 2, 3, 4, 5, 6, 7</span>
<span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="nf">delete_intervals</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">6000</span><span class="p">,</span><span class="w"> </span><span class="m">10000</span><span class="p">)))</span><span class="w">  </span><span class="c1"># delete data after 6kb</span>
<span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="nf">trim</span><span class="p">()</span><span class="w">  </span><span class="c1"># remove the deleted region</span>
<span class="nf">paste</span><span class="p">(</span>
<span class="w">    </span><span class="s">&quot;Reduced from&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">num_trees</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;trees over&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">sequence_length</span><span class="o">/</span><span class="m">1e3</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;kb to&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">reduced_ts</span><span class="o">$</span><span class="n">num_trees</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;trees over&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="n">sequence_length</span><span class="o">/</span><span class="m">1e3</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;kb.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Reduced from 26 trees over 10 kb to 6 trees over 6 kb.'</div></div>
</div>
</section>
<section id="ids-and-indexes">
<h2>IDs and indexes<a class="headerlink" href="#ids-and-indexes" title="Link to this heading">#</a></h2>
<p>Note that if a bare digit is provided to one of these methods, it will be treated as a
floating point number (numeric in R). This is useful to know when calling <code class="docutils literal notranslate"><span class="pre">tskit</span></code> methods that
require integers (such as object IDs). For example, the following will not work:</p>
<div class="cell tag_raises-exception tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="m">0</span><span class="p">)</span><span class="w">  </span><span class="c1"># Will raise a TypeError</span>
<span class="nf">is</span><span class="p">(</span><span class="m">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In this case, to force the <code class="docutils literal notranslate"><span class="pre">0</span></code> to be passed as an integer, you can either coerce it
using <code class="docutils literal notranslate"><span class="pre">as.integer</span></code> or simply append the letter <code class="docutils literal notranslate"><span class="pre">L</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">is</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="m">0</span><span class="p">))</span>
<span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="m">0</span><span class="p">))</span>
<span class="c1"># or</span>
<span class="nf">is</span><span class="p">(</span><span class="m">0L</span><span class="p">)</span>
<span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="m">0L</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'integer'</li><li>'double'</li><li>'numeric'</li><li>'vector'</li><li>'data.frameRowLabels'</li><li>'index'</li><li>'replValue'</li><li>'numLike'</li><li>'number'</li><li>'atomicVector'</li><li>'replValueSp'</li></ol>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=0, flags=1, time=0.0, population=0, individual=0, metadata=b&#39;&#39;)
</pre></div>
</div>
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'integer'</li><li>'double'</li><li>'numeric'</li><li>'vector'</li><li>'data.frameRowLabels'</li><li>'index'</li><li>'replValue'</li><li>'numLike'</li><li>'number'</li><li>'atomicVector'</li><li>'replValueSp'</li></ol>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=0, flags=1, time=0.0, population=0, individual=0, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Coercing in this way is only necessary when passing parameters to those underlying
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> methods that expect integers. It is not needed to index into numeric arrays.
<em>However</em>, when using arrays, very careful attention must be paid to the fact that
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> IDs start at zero, whereas R indexes start at one:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">root_id_int</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">first</span><span class="p">()</span><span class="o">$</span><span class="n">root</span>
<span class="nf">is</span><span class="p">(</span><span class="n">root_id_int</span><span class="p">)</span>
<span class="n">root_id_num</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">root_id_int</span><span class="p">)</span>

<span class="c1"># Using integer ID will work</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Root time via tskit method:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="n">root_id_int</span><span class="p">)</span><span class="o">$</span><span class="n">time</span><span class="p">)</span>
<span class="c1"># Using numeric ID will not work</span>
<span class="c1"># paste(&quot;Root time via tskit method:&quot;, ts$node(root_id_num)$time)</span>

<span class="c1"># When indexing into tskit arrays in R, add 1 to the ID (integer or numeric)</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Root time via array access:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">nodes_time</span><span class="p">[</span><span class="n">root_id_int</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">])</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Root time via array access:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">nodes_time</span><span class="p">[</span><span class="n">root_id_num</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'integer'</li><li>'double'</li><li>'numeric'</li><li>'vector'</li><li>'data.frameRowLabels'</li><li>'index'</li><li>'replValue'</li><li>'numLike'</li><li>'number'</li><li>'atomicVector'</li><li>'replValueSp'</li></ol>
</div><div class="output text_html">'Root time via tskit method: 3.4634393266485'</div><div class="output text_html">'Root time via array access: 3.4634393266485'</div><div class="output text_html">'Root time via array access: 3.4634393266485'</div></div>
</div>
</section>
<section id="analysis">
<h2>Analysis<a class="headerlink" href="#analysis" title="Link to this heading">#</a></h2>
<p>From within R we can use <code class="docutils literal notranslate"><span class="pre">tskit</span></code>’s powerful
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/stats.html">Statistics</a> framework to efficiently
compute many different summary statistics from a tree sequence. To illustrate this,
we’ll first add some mutations to our tree sequence with the
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_mutations" title="(in Project name not set)"><code class="docutils literal notranslate"><span class="pre">msprime.sim_mutations()</span></code></a> function, and then compute the genetic diversity
for each of the tree sequence’s sample nodes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">msprime</span><span class="o">$</span><span class="nf">sim_mutations</span><span class="p">(</span><span class="n">reduced_ts</span><span class="p">,</span><span class="w"> </span><span class="n">rate</span><span class="o">=</span><span class="m">1e-4</span><span class="p">,</span><span class="w"> </span><span class="n">random_seed</span><span class="o">=</span><span class="m">321</span><span class="p">)</span>
<span class="nf">paste</span><span class="p">(</span><span class="n">ts_mut</span><span class="o">$</span><span class="n">num_mutations</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;mutations, genetic diversity is&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">diversity</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'13 mutations, genetic diversity is 0.000720238095238095'</div></div>
</div>
<p>Numerical arrays and matrices work as expected. For instance, we can use the tree
sequence <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.genotype_matrix" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genotype_matrix()</span></code></a> method to return the genotypes of
the tree sequence as a matrix object in R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">G</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">genotype_matrix</span><span class="p">()</span>
<span class="n">G</span>
<span class="nf">is</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 13 × 8 of type int</caption>
<tbody>
	<tr><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>1</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td></tr>
</tbody>
</table>
</div><div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'matrix'</li><li>'array'</li><li>'replValueSp'</li><li>'structure'</li><li>'vector'</li></ol>
</div></div>
</div>
<p>We can then use R functions directly on the genotype matrix:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">allele_frequency</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowMeans</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
<span class="n">allele_frequency</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>0.125</li><li>0.125</li><li>0.125</li><li>0.125</li><li>0.5</li><li>0.625</li><li>0.875</li><li>0.875</li><li>0.125</li><li>0.25</li><li>0.125</li><li>0.625</li><li>0.125</li></ol>
</div></div>
</div>
</section>
<section id="jupyter-notebook-tips">
<h2>Jupyter notebook tips<a class="headerlink" href="#jupyter-notebook-tips" title="Link to this heading">#</a></h2>
<p>When running R within a <a class="reference external" href="https://jupyter.org">Jupyter notebook</a>, a few magic functions
can be defined that allow tskit objects to be rendered within the notebook:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define some magic functions to allow objects to be displayed in R Jupyter notebooks</span>
<span class="n">repr_html.tskit.trees.TreeSequence</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`_repr_html_`</span><span class="p">()}</span>
<span class="n">repr_html.tskit.trees.Tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`_repr_html_`</span><span class="p">()}</span>
<span class="n">repr_svg.tskit.drawing.SVGString</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`__str__`</span><span class="p">()}</span>
</pre></div>
</div>
</div>
</div>
<p>This leads to much nicer tabular summaries:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>6</td></tr>
                      <tr><td>Sequence Length</td><td>6000.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>8</td></tr>
                      <tr><td>Total Size</td><td>6.1 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>32</td>
                      <td>1.0 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>4</td>
                      <td>136 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>13</td>
                      <td>497 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>20</td>
                      <td>568 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>5</td>
                      <td>3.1 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>13</td>
                      <td>341 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>It also allows trees and tree sequences to be plotted inline:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">y_ticks</span><span class="o">=</span><span class="m">0</span><span class="o">:</span><span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'&lt;svg baseProfile="full" height="200" version="1.1" width="1200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;&lt;defs&gt;&lt;style type="text/css"&gt;&lt;![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node &gt; .sym {fill: black; stroke: none}.site &gt; .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]&gt;&lt;/style&gt;&lt;/defs&gt;&lt;g class="tree-sequence"&gt;&lt;g class="background"&gt;&lt;path d="M56.8,0 l187.2,0 l0,138.2 l-144.144,25 l0,5 l-43.056,0 l0,-5 l0,-25 l0,-138.2z" /&gt;&lt;path d="M244,0 l187.2,0 l0,138.2 l123.739,25 l0,5 l-455.083,0 l0,-5 l144.144,-25 l0,-138.2z" /&gt;&lt;path d="M431.2,0 l187.2,0 l0,138.2 l217.714,25 l0,5 l-281.174,0 l0,-5 l-123.739,-25 l0,-138.2z" /&gt;&lt;path d="M618.4,0 l187.2,0 l0,138.2 l184.954,25 l0,5 l-154.44,0 l0,-5 l-217.714,-25 l0,-138.2z" /&gt;&lt;path d="M805.6,0 l187.2,0 l0,138.2 l129.168,25 l0,5 l-131.414,0 l0,-5 l-184.954,-25 l0,-138.2z" /&gt;&lt;path d="M992.8,0 l187.2,0 l0,138.2 l0,25 l0,5 l-58.032,0 l0,-5 l-129.17,-25 l0,-138.2z" /&gt;&lt;/g&gt;&lt;g class="axes"&gt;&lt;g class="x-axis"&gt;&lt;g class="title" transform="translate(618.4 200)"&gt;&lt;text class="lab" text-anchor="middle" transform="translate(0 -11)"&gt;Genome position&lt;/text&gt;&lt;/g&gt;&lt;line class="ax-line" x1="56.8" x2="1180" y1="163.2" y2="163.2" /&gt;&lt;g class="ticks"&gt;&lt;g class="tick" transform="translate(56.8 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(99.856 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;230&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(554.939 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;2661&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(836.114 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;4163&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(990.554 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;4988&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(1121.97 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;5690&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(1180 163.2)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5" /&gt;&lt;g transform="translate(0 6)"&gt;&lt;text class="lab"&gt;6000&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s0" transform="translate(144.222 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m0"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s1" transform="translate(334.979 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m1"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s2" transform="translate(413.603 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m2"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s3" transform="translate(433.259 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m3"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s4" transform="translate(663.702 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m4"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s5" transform="translate(749.627 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m5"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s6" transform="translate(842.478 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m6"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s7" transform="translate(878.795 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m7"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s8" transform="translate(916.984 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m8"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s9" transform="translate(934.955 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m9"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s10" transform="translate(1059.63 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m10"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s11" transform="translate(1090.71 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m11"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="site s12" transform="translate(1107.37 163.2)"&gt;&lt;line class="sym" x1="0" x2="0" y1="0" y2="-10" /&gt;&lt;g class="mut m12"&gt;&lt;polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5" /&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="y-axis"&gt;&lt;g class="title" transform="translate(0 65.7)"&gt;&lt;text class="lab" text-anchor="middle" transform="translate(11) rotate(-90)"&gt;Time (generations)&lt;/text&gt;&lt;/g&gt;&lt;line class="ax-line" x1="56.8" x2="56.8" y1="121.4" y2="10" /&gt;&lt;g class="ticks"&gt;&lt;g class="tick" transform="translate(56.8 121.4)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 108.736)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 96.073)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 83.4095)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 70.746)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 58.0824)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 45.4189)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 32.7554)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 20.0919)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 7.42839)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tick" transform="translate(56.8 -5.23512)"&gt;&lt;line x1="0" x2="-5" y1="0" y2="0" /&gt;&lt;g transform="translate(-6 0)"&gt;&lt;text class="lab" text-anchor="end"&gt;10&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="plotbox trees"&gt;&lt;g class="tree t0" transform="translate(56.8 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n18 p0 root" transform="translate(63.9875 77.5407)"&gt;&lt;g class="a18 c2 node n13 p0" transform="translate(25.5875 30.7712)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-23.575 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 23.575" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 node n12 p0" transform="translate(23.575 2.03424)"&gt;&lt;g class="a12 i1 leaf node n2 p0 sample" transform="translate(-28.75 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 28.75" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n10 p0" transform="translate(28.75 5.24241)"&gt;&lt;g class="a10 i2 leaf node n4 p0 sample" transform="translate(16.1 5.81144)"&gt;&lt;path class="edge" d="M 0 0 V -5.81144 H -16.1" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a10 c2 node n9 p0" transform="translate(-16.1 1.39353)"&gt;&lt;g class="a9 i2 leaf node n5 p0 sample" transform="translate(13.8 4.41791)"&gt;&lt;path class="edge" d="M 0 0 V -4.41791 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a9 c2 node n8 p0" transform="translate(-13.8 0.8617)"&gt;&lt;g class="a8 i1 leaf node n3 p0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf node n6 p0 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -0.8617 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -1.39353 H 16.1" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -5.24241 H -28.75" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;10&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -23.575" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -30.7712 H -25.5875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;g class="a18 c2 node n14 p0" transform="translate(-25.5875 21.7136)"&gt;&lt;g class="a14 i0 leaf node n0 p0 sample" transform="translate(-9.2 22.1457)"&gt;&lt;path class="edge" d="M 0 0 V -22.1457 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a14 i3 leaf node n7 p0 sample" transform="translate(9.2 22.1457)"&gt;&lt;path class="edge" d="M 0 0 V -22.1457 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -21.7136 H 25.5875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;14&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;18&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tree t1" transform="translate(244 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n18 p0 root" transform="translate(63.9875 77.5407)"&gt;&lt;g class="a18 c2 node n13 p0" transform="translate(25.5875 30.7712)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-23.575 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 23.575" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 node n12 p0" transform="translate(23.575 2.03424)"&gt;&lt;g class="a12 i1 leaf m3 node n2 p0 s3 sample" transform="translate(-28.75 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 28.75" /&gt;&lt;g class="mut m3 s3" transform="translate(0 -3.79834)"&gt;&lt;line x1="0" x2="0" y1="0" y2="3.79834" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n10 p0" transform="translate(28.75 5.24241)"&gt;&lt;g class="a10 i2 leaf node n4 p0 sample" transform="translate(16.1 5.81144)"&gt;&lt;path class="edge" d="M 0 0 V -5.81144 H -16.1" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a10 c2 node n9 p0" transform="translate(-16.1 1.39353)"&gt;&lt;g class="a9 i2 leaf node n5 p0 sample" transform="translate(13.8 4.41791)"&gt;&lt;path class="edge" d="M 0 0 V -4.41791 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a9 c2 node n8 p0" transform="translate(-13.8 0.8617)"&gt;&lt;g class="a8 i1 leaf m0 node n3 p0 s0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;g class="mut m0 s0" transform="translate(0 -2.73828)"&gt;&lt;line x1="0" x2="0" y1="0" y2="2.73828" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf m1 node n6 p0 s1 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;g class="mut m1 s1" transform="translate(0 -1.38406)"&gt;&lt;line x1="0" x2="0" y1="0" y2="1.38406" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -0.8617 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -1.39353 H 16.1" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -5.24241 H -28.75" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;10&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -23.575" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -30.7712 H -25.5875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;g class="a18 c2 node n16 p0" transform="translate(-25.5875 6.71959)"&gt;&lt;g class="a16 i0 leaf m2 node n0 p0 s2 sample" transform="translate(-9.2 37.1397)"&gt;&lt;path class="edge" d="M 0 0 V -37.1397 H 9.2" /&gt;&lt;g class="mut m2 s2" transform="translate(0 -29.5048)"&gt;&lt;line x1="0" x2="0" y1="0" y2="29.5048" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a16 i3 leaf node n7 p0 sample" transform="translate(9.2 37.1397)"&gt;&lt;path class="edge" d="M 0 0 V -37.1397 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -6.71959 H 25.5875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;16&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;18&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tree t2" transform="translate(431.2 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n18 p0 root" transform="translate(73.475 77.5407)"&gt;&lt;g class="a18 c2 m5 node n13 p0 s5" transform="translate(30.475 30.7712)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-19.55 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 19.55" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 m4 node n12 p0 s4" transform="translate(19.55 2.03424)"&gt;&lt;g class="a12 i1 leaf node n2 p0 sample" transform="translate(-20.7 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 20.7" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n9 p0" transform="translate(20.7 6.63594)"&gt;&lt;g class="a9 i2 leaf node n5 p0 sample" transform="translate(13.8 4.41791)"&gt;&lt;path class="edge" d="M 0 0 V -4.41791 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a9 c2 node n8 p0" transform="translate(-13.8 0.8617)"&gt;&lt;g class="a8 i1 leaf node n3 p0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf node n6 p0 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -0.8617 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -6.63594 H -20.7" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -19.55" /&gt;&lt;g class="mut m4 s4" transform="translate(0 -1.54933)"&gt;&lt;line x1="0" x2="0" y1="0" y2="1.54933" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -30.7712 H -30.475" /&gt;&lt;g class="mut m5 s5" transform="translate(0 -28.0667)"&gt;&lt;line x1="0" x2="0" y1="0" y2="28.0667" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;g class="a18 c2 node n16 p0" transform="translate(-30.475 6.71959)"&gt;&lt;g class="a16 i0 leaf node n0 p0 sample" transform="translate(-13.8 37.1397)"&gt;&lt;path class="edge" d="M 0 0 V -37.1397 H 13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a16 c2 node n11 p0" transform="translate(13.8 29.6414)"&gt;&lt;g class="a11 i2 leaf node n4 p0 sample" transform="translate(-9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a11 i3 leaf node n7 p0 sample" transform="translate(9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -29.6414 H -13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;11&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -6.71959 H 30.475" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;16&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;18&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tree t3" transform="translate(618.4 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n19 p0 root" transform="translate(68.5875 26.8)"&gt;&lt;g class="a19 i0 leaf m8 node n0 p0 s8 sample" transform="translate(-39.3875 94.6)"&gt;&lt;path class="edge" d="M 0 0 V -94.6 H 39.3875" /&gt;&lt;g class="mut m8 s8" transform="translate(0 -30.0795)"&gt;&lt;line x1="0" x2="0" y1="0" y2="30.0795" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a19 c2 m6 m7 node n18 p0 s6 s7" transform="translate(39.3875 50.7407)"&gt;&lt;g class="a18 c2 node n13 p0" transform="translate(-40.825 30.7712)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-19.55 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 19.55" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 node n12 p0" transform="translate(19.55 2.03424)"&gt;&lt;g class="a12 i1 leaf node n2 p0 sample" transform="translate(-20.7 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 20.7" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n9 p0" transform="translate(20.7 6.63594)"&gt;&lt;g class="a9 i2 leaf node n5 p0 sample" transform="translate(13.8 4.41791)"&gt;&lt;path class="edge" d="M 0 0 V -4.41791 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a9 c2 node n8 p0" transform="translate(-13.8 0.8617)"&gt;&lt;g class="a8 i1 leaf node n3 p0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf node n6 p0 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -0.8617 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -6.63594 H -20.7" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -19.55" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -30.7712 H 40.825" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;g class="a18 c2 m9 node n11 p0 s9" transform="translate(40.825 36.361)"&gt;&lt;g class="a11 i2 leaf node n4 p0 sample" transform="translate(-9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a11 i3 leaf node n7 p0 sample" transform="translate(9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -36.361 H -40.825" /&gt;&lt;g class="mut m9 s9" transform="translate(0 -5.53803)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5.53803" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;11&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -50.7407 H -39.3875" /&gt;&lt;g class="mut m7 s7" transform="translate(0 -45.4606)"&gt;&lt;line x1="0" x2="0" y1="0" y2="45.4606" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;g class="mut m6 s6" transform="translate(0 -7.38293)"&gt;&lt;line x1="0" x2="0" y1="0" y2="7.38293" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;18&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;19&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tree t4" transform="translate(805.6 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n19 p0 root" transform="translate(98.4875 26.8)"&gt;&lt;g class="a19 c2 node n11 p0" transform="translate(50.3125 87.1017)"&gt;&lt;g class="a11 i2 leaf m12 node n4 p0 s12 sample" transform="translate(-9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H 9.2" /&gt;&lt;g class="mut m12 s12" transform="translate(0 -5.04465)"&gt;&lt;line x1="0" x2="0" y1="0" y2="5.04465" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a11 i3 leaf node n7 p0 sample" transform="translate(9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -87.1017 H -50.3125" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;11&lt;/text&gt;&lt;/g&gt;&lt;g class="a19 c2 node n17 p0" transform="translate(-50.3125 51.63)"&gt;&lt;g class="a17 i0 leaf m10 node n0 p0 s10 sample" transform="translate(-18.975 42.97)"&gt;&lt;path class="edge" d="M 0 0 V -42.97 H 18.975" /&gt;&lt;g class="mut m10 s10" transform="translate(0 -10.9638)"&gt;&lt;line x1="0" x2="0" y1="0" y2="10.9638" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab lft" transform="translate(-5 0)"&gt;10&lt;/text&gt;&lt;/g&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a17 c2 m11 node n13 p0 s11" transform="translate(18.975 29.8819)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-19.55 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 19.55" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 node n12 p0" transform="translate(19.55 2.03424)"&gt;&lt;g class="a12 i1 leaf node n2 p0 sample" transform="translate(-20.7 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 20.7" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n9 p0" transform="translate(20.7 6.63594)"&gt;&lt;g class="a9 i2 leaf node n5 p0 sample" transform="translate(13.8 4.41791)"&gt;&lt;path class="edge" d="M 0 0 V -4.41791 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a9 c2 node n8 p0" transform="translate(-13.8 0.8617)"&gt;&lt;g class="a8 i1 leaf node n3 p0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf node n6 p0 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -0.8617 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -6.63594 H -20.7" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;9&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -19.55" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -29.8819 H -18.975" /&gt;&lt;g class="mut m11 s11" transform="translate(0 -24.632)"&gt;&lt;line x1="0" x2="0" y1="0" y2="24.632" /&gt;&lt;path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6" /&gt;&lt;text class="lab rgt" transform="translate(5 0)"&gt;11&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -51.63 H 50.3125" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;17&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;19&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;g class="tree t5" transform="translate(992.8 0)"&gt;&lt;g class="plotbox"&gt;&lt;g class="c2 node n19 p0 root" transform="translate(95.325 26.8)"&gt;&lt;g class="a19 c2 node n17 p0" transform="translate(-48.875 51.63)"&gt;&lt;g class="a17 i0 leaf node n0 p0 sample" transform="translate(-17.25 42.97)"&gt;&lt;path class="edge" d="M 0 0 V -42.97 H 17.25" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;0&lt;/text&gt;&lt;/g&gt;&lt;g class="a17 c2 node n13 p0" transform="translate(17.25 29.8819)"&gt;&lt;g class="a13 i0 leaf node n1 p0 sample" transform="translate(-16.1 13.0881)"&gt;&lt;path class="edge" d="M 0 0 V -13.0881 H 16.1" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;1&lt;/text&gt;&lt;/g&gt;&lt;g class="a13 c2 node n12 p0" transform="translate(16.1 2.03424)"&gt;&lt;g class="a12 i1 leaf node n2 p0 sample" transform="translate(-13.8 11.0538)"&gt;&lt;path class="edge" d="M 0 0 V -11.0538 H 13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;2&lt;/text&gt;&lt;/g&gt;&lt;g class="a12 c2 node n8 p0" transform="translate(13.8 7.49764)"&gt;&lt;g class="a8 i1 leaf node n3 p0 sample" transform="translate(-9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;3&lt;/text&gt;&lt;/g&gt;&lt;g class="a8 i3 leaf node n6 p0 sample" transform="translate(9.2 3.55621)"&gt;&lt;path class="edge" d="M 0 0 V -3.55621 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;6&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -7.49764 H -13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;8&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -2.03424 H -16.1" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;12&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -29.8819 H -17.25" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;13&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -51.63 H 48.875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;17&lt;/text&gt;&lt;/g&gt;&lt;g class="a19 c2 node n15 p0" transform="translate(48.875 57.9197)"&gt;&lt;g class="a15 i2 leaf node n5 p0 sample" transform="translate(13.8 36.6803)"&gt;&lt;path class="edge" d="M 0 0 V -36.6803 H -13.8" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;5&lt;/text&gt;&lt;/g&gt;&lt;g class="a15 c2 node n11 p0" transform="translate(-13.8 29.182)"&gt;&lt;g class="a11 i2 leaf node n4 p0 sample" transform="translate(-9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H 9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;4&lt;/text&gt;&lt;/g&gt;&lt;g class="a11 i3 leaf node n7 p0 sample" transform="translate(9.2 7.49833)"&gt;&lt;path class="edge" d="M 0 0 V -7.49833 H -9.2" /&gt;&lt;rect class="sym" height="6" width="6" x="-3" y="-3" /&gt;&lt;text class="lab" transform="translate(0 11)"&gt;7&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -29.182 H 13.8" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab lft" transform="translate(-3 -7.0)"&gt;11&lt;/text&gt;&lt;/g&gt;&lt;path class="edge" d="M 0 0 V -57.9197 H -48.875" /&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab rgt" transform="translate(3 -7.0)"&gt;15&lt;/text&gt;&lt;/g&gt;&lt;circle class="sym" cx="0" cy="0" r="3" /&gt;&lt;text class="lab" transform="translate(0 -11)"&gt;19&lt;/text&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;/g&gt;&lt;/svg&gt;'</div></div>
</div>
</section>
<section id="interaction-with-r-libraries">
<h2>Interaction with R libraries<a class="headerlink" href="#interaction-with-r-libraries" title="Link to this heading">#</a></h2>
<p>R has a number of libraries to deal with genomic data and trees. Below we demo the
phylogenetic tree representation defined in the the popular
<a class="reference external" href="http://ape-package.ird.fr">ape</a> package, taking all the trees
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.write_nexus" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">exported</span> <span class="pre">in</span> <span class="pre">Nexus</span> <span class="pre">format</span></code></a>, or
individual trees <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.as_newick" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">exported</span> <span class="pre">in</span> <span class="pre">Newick</span> <span class="pre">format</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tempfile</span><span class="p">()</span>
<span class="n">ts_mut</span><span class="o">$</span><span class="nf">write_nexus</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="c1"># Warning - ape trees are stored independently, so this will use much more memory than tskit</span>
<span class="n">trees</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ape</span><span class="o">::</span><span class="nf">read.nexus</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">force.multi</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">  </span><span class="c1"># return a set of trees</span>

<span class="c1"># Or simply read in a single tree</span>
<span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ape</span><span class="o">::</span><span class="nf">read.tree</span><span class="p">(</span><span class="n">text</span><span class="o">=</span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">first</span><span class="p">()</span><span class="o">$</span><span class="nf">as_newick</span><span class="p">())</span>

<span class="c1"># Now we can plot the tree in tskit style, but using the ape library</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">direction</span><span class="o">=</span><span class="s">&quot;downward&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">srt</span><span class="o">=</span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">adj</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span><span class="w">  </span><span class="c1"># or equivalently use trees[[1]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/ba4e21555374a0a3002f3cbfaba7386fcc049d206857d1b97cda5bae4dc2c46e.png"><img alt="_images/ba4e21555374a0a3002f3cbfaba7386fcc049d206857d1b97cda5bae4dc2c46e.png" src="_images/ba4e21555374a0a3002f3cbfaba7386fcc049d206857d1b97cda5bae4dc2c46e.png" style="width: 420px; height: 420px;" /></a>
</div>
</div>
<p>Note that nodes are labelled with the prefix <code class="docutils literal notranslate"><span class="pre">n</span></code>, so that nodes <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, …
become <code class="docutils literal notranslate"><span class="pre">n0</span></code>, <code class="docutils literal notranslate"><span class="pre">n1</span></code>, <code class="docutils literal notranslate"><span class="pre">n2</span></code>, … This helps to avoid
confusion between the the zero-based counting system used natively
by <code class="docutils literal notranslate"><span class="pre">tskit</span></code>, and the one-based counting system used in <code class="docutils literal notranslate"><span class="pre">R</span></code>.</p>
</section>
<section id="further-information">
<h2>Further information<a class="headerlink" href="#further-information" title="Link to this heading">#</a></h2>
<p>Be sure to check out the <a class="reference external" href="https://rstudio.github.io/reticulate">reticulate</a>
documentation, in particular on
<a class="reference external" href="https://rstudio.github.io/reticulate/articles/calling_python.html">Calling Python from R</a>,
which includes important information on how R data types are converted to their
equivalent Python types.</p>
</section>
<section id="advanced-usage-through-the-tskit-s-c-api">
<h2>Advanced usage through the tskit’s C API<a class="headerlink" href="#advanced-usage-through-the-tskit-s-c-api" title="Link to this heading">#</a></h2>
<p>While the approach with <code class="docutils literal notranslate"><span class="pre">reticulate</span></code> is very powerful and will be useful for
most analyses, it does have an overhead due to calling Python and conversion of objects.
This overhead will be minimal for some use cases but could be significant for others.
An alternative approach is to use the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> C API from R,
via the standard <a class="reference external" href="https://cran.r-project.org/doc/manuals/R-exts.html">R Extensions</a> or
via <a class="reference external" href="https://www.rcpp.org">Rcpp</a>,
but these are advanced topics beyond the scope of this tutorial.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="more_forward_sims.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><em>Advanced forward simulations</em></p>
      </div>
    </a>
    <a class="right-next"
       href="tutorial_development.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><em>Developing new tutorials</em></p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#attributes-and-methods">Attributes and methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ids-and-indexes">IDs and indexes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis">Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#jupyter-notebook-tips">Jupyter notebook tips</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interaction-with-r-libraries">Interaction with R libraries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-information">Further information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#advanced-usage-through-the-tskit-s-c-api">Advanced usage through the tskit’s C API</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>