
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Tskit and R &#8212; Tree Sequence Tutorials</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/tskit-book-theme.css?digest=4e121f05f62280197eeb3e8760a1903920f0778f" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/tskit-book-theme.js?digest=9a6366feaade5ec470560844be9ff19788eb9b85"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Developing new tutorials" href="tutorial_development.html" />
    <link rel="prev" title="Building a forward simulator" href="forward_sims.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/tskit_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Tree Sequence Tutorials</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome!
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="what_is.html">
   What is a tree sequence?
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tree sequence basics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="terminology_and_concepts.html">
   Terminology &amp; concepts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="getting_started.html">
   Getting started with
   <strong class="program">
    tskit
   </strong>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistics and Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="analysing_tree_sequences.html">
   <em>
    Analysing tree sequences
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysing_trees.html">
   Analysing trees
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="counting_topologies.html">
   <em>
    Counting topologies
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="parallelization.html">
   <em>
    Parallelization
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Further tskit tutorials
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tables_and_editing.html">
   Tables and editing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="simplification.html">
   <em>
    Simplification
   </em>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="viz.html">
   Visualization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   Working with Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="args.html">
   <em>
    ARGs as tree sequences
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Simulation
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="simulation_overview.html">
   Tree sequences and simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="no_mutations.html">
   Do you really need mutations?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="completing_forward_sims.html">
   Completing forwards simulations
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="advanced_msprime.html">
   Advanced
   <strong class="program">
    msprime
   </strong>
   topics
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="demography.html">
     Demography
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="bottlenecks.html">
     Instantaneous Bottlenecks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="introgression.html">
     Introgression
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="forward_sims.html">
   <em>
    Building a forward simulator
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Other languages
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Tskit and R
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Development
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial_development.html">
   <em>
    Developing new tutorials
   </em>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tskit for ...
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="popgen.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     Tskit
    </span>
   </code>
   for population genetics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="phylogen.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     Tskit
    </span>
   </code>
   for phylogenetics
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/tskit-dev/tutorials"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/tskit-dev/tutorials/issues/new?title=Issue%20on%20page%20%2Ftskitr.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/tskitr.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download notebook file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        <a href="_sources/tskitr.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#attributes-and-methods">
   Attributes and methods
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ids-and-indexes">
     IDs and indexes
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis">
   Analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#jupyter-notebook-tips">
   Jupyter notebook tips
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#interaction-with-r-libraries">
   Interaction with R libraries
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#further-information">
   Further information
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Tskit and R</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#attributes-and-methods">
   Attributes and methods
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ids-and-indexes">
     IDs and indexes
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analysis">
   Analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#jupyter-notebook-tips">
   Jupyter notebook tips
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#interaction-with-r-libraries">
   Interaction with R libraries
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#further-information">
   Further information
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="tskit-and-r">
<span id="sec-tskit-r"></span><h1>Tskit and R<a class="headerlink" href="#tskit-and-r" title="Permalink to this headline">#</a></h1>
<p>To interface with <code class="docutils literal notranslate"><span class="pre">tskit</span></code> in R, we can use the <a class="reference external" href="https://rstudio.github.io/reticulate/">reticulate</a> R package, which lets you call Python functions within an R session. In this tutorial, we’ll go through a couple of examples to show you how to get started. If you haven’t done so already, you’ll need to install <code class="docutils literal notranslate"><span class="pre">reticulate</span></code> in your R session via <code class="docutils literal notranslate"><span class="pre">install.packages(&quot;reticulate&quot;)</span></code>.</p>
<p>We’ll begin by simulating a small tree sequence using <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">msprime</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reticulate</span><span class="o">::</span><span class="nf">import</span><span class="p">(</span><span class="s">&quot;msprime&quot;</span><span class="p">)</span>

<span class="n">ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">msprime</span><span class="o">$</span><span class="nf">sim_ancestry</span><span class="p">(</span><span class="m">80</span><span class="p">,</span><span class="w"> </span><span class="n">sequence_length</span><span class="o">=</span><span class="m">1e4</span><span class="p">,</span><span class="w"> </span><span class="n">recombination_rate</span><span class="o">=</span><span class="m">1e-4</span><span class="p">,</span><span class="w"> </span><span class="n">random_seed</span><span class="o">=</span><span class="m">42</span><span class="p">)</span>
<span class="n">ts</span><span class="w">  </span><span class="c1"># See &quot;Jupyter notebook tips&quot;, below for how to render this nicely</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │         26║
╟───────────────┼───────────╢
║Sequence Length│      10000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │        160║
╟───────────────┼───────────╢
║Total Size     │   29.0 KiB║
╚═══════════════╧═══════════╝
╔═══════════╤════╤═════════╤════════════╗
║Table      │Rows│Size     │Has Metadata║
╠═══════════╪════╪═════════╪════════════╣
║Edges      │ 414│ 12.9 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Individuals│  80│  2.2 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Migrations │   0│  8 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Mutations  │   0│ 16 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Nodes      │ 344│  9.4 KiB│          No║
╟───────────┼────┼─────────┼────────────╢
║Populations│   1│224 Bytes│         Yes║
╟───────────┼────┼─────────┼────────────╢
║Provenances│   1│948 Bytes│          No║
╟───────────┼────┼─────────┼────────────╢
║Sites      │   0│ 16 Bytes│          No║
╚═══════════╧════╧═════════╧════════════╝
</pre></div>
</div>
</div>
</div>
<section id="attributes-and-methods">
<h2>Attributes and methods<a class="headerlink" href="#attributes-and-methods" title="Permalink to this headline">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">reticulate</span></code> allows us to access a Python object’s attributes via
the <code class="docutils literal notranslate"><span class="pre">$</span></code> operator. For example, we can access (and assign to a variable) the number of
samples in the tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">num_samples</span>
<span class="n">n</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">160</div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">$</span></code> operator can also be used to call methods, for example, the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> method associated with the tree sequence.
The method parameters are given as native R objects
(but note that object IDs still use tskit’s 0-based indexing system).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">simplify</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">7</span><span class="p">)</span><span class="w">  </span><span class="c1"># only keep samples with ids 0, 1, 2, 3, 4, 5, 6, 7</span>
<span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="nf">delete_intervals</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">6000</span><span class="p">,</span><span class="w"> </span><span class="m">10000</span><span class="p">)))</span><span class="w">  </span><span class="c1"># delete data after 6kb</span>
<span class="n">reduced_ts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="nf">trim</span><span class="p">()</span><span class="w">  </span><span class="c1"># remove the deleted region</span>
<span class="nf">paste</span><span class="p">(</span>
<span class="w">    </span><span class="s">&quot;Reduced from&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">num_trees</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;trees over&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">sequence_length</span><span class="o">/</span><span class="m">1e3</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;kb to&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">reduced_ts</span><span class="o">$</span><span class="n">num_trees</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;trees over&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">reduced_ts</span><span class="o">$</span><span class="n">sequence_length</span><span class="o">/</span><span class="m">1e3</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;kb.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Reduced from 26 trees over 10 kb to 6 trees over 6 kb.'</div></div>
</div>
<section id="ids-and-indexes">
<h3>IDs and indexes<a class="headerlink" href="#ids-and-indexes" title="Permalink to this headline">#</a></h3>
<p>Note that if a bare digit is provided to one of these methods, it will be treated as a
floating point number. This is useful to know when calling <code class="docutils literal notranslate"><span class="pre">tskit</span></code> methods that
require integers (e.g. object IDs). For example, the following will not work:</p>
<div class="cell tag_raises-exception tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="m">0</span><span class="p">)</span><span class="w">  </span><span class="c1"># Will raise an error</span>
</pre></div>
</div>
</div>
</div>
<p>In this case, to force the <code class="docutils literal notranslate"><span class="pre">0</span></code> to be passed as an integer, you can either coerce it
using <code class="docutils literal notranslate"><span class="pre">as.integer</span></code> or simply prepend the letter <code class="docutils literal notranslate"><span class="pre">L</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="m">0</span><span class="p">))</span>
<span class="c1"># or</span>
<span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="m">0L</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=0, flags=1, time=0.0, population=0, individual=0, metadata=b&#39;&#39;)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=0, flags=1, time=0.0, population=0, individual=0, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Coercing in this way is only necessary when passing parameters to those underlying
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> methods that expect integers. It is not needed e.g. to index into numeric arrays.
<em>However</em>, when using arrays, very careful attention must be paid to the fact that
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> IDs start at zero, whereas R indexes start at one:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">root_id</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">first</span><span class="p">()</span><span class="o">$</span><span class="n">root</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Root time via tskit method:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="nf">node</span><span class="p">(</span><span class="n">root_id</span><span class="p">)</span><span class="o">$</span><span class="n">time</span><span class="p">)</span>
<span class="c1"># When indexing into tskit arrays in R, add 1 to the ID</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Root time via array access:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts</span><span class="o">$</span><span class="n">nodes_time</span><span class="p">[</span><span class="n">root_id</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Root time via tskit method: 3.4634393266485'</div><div class="output text_html">'Root time via array access: 3.4634393266485'</div></div>
</div>
</section>
</section>
<section id="analysis">
<h2>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">#</a></h2>
<p>From within R we can use <code class="docutils literal notranslate"><span class="pre">tskit</span></code>’s powerful
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/stats.html">Statistics</a> framework to efficiently
compute many different summary statistics from a tree sequence. To illustrate this,
we’ll first add some mutations to our tree sequence with the
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_mutations" title="(in Python)"><code class="docutils literal notranslate"><span class="pre">msprime.sim_mutations()</span></code></a> function, and then compute the genetic diversity
for each of the tree sequence’s sample nodes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">msprime</span><span class="o">$</span><span class="nf">sim_mutations</span><span class="p">(</span><span class="n">reduced_ts</span><span class="p">,</span><span class="w"> </span><span class="n">rate</span><span class="o">=</span><span class="m">1e-4</span><span class="p">,</span><span class="w"> </span><span class="n">random_seed</span><span class="o">=</span><span class="m">321</span><span class="p">)</span>

<span class="nf">paste</span><span class="p">(</span><span class="n">ts_mut</span><span class="o">$</span><span class="n">num_mutations</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;mutations, genetic diversity is&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">diversity</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'13 mutations, genetic diversity is 0.000720238095238095'</div></div>
</div>
<p>Numerical arrays and matrices work as expected. For instance, we can use the tree
sequence <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.genotype_matrix" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genotype_matrix()</span></code></a> method to return the genotypes of
the tree sequence as a matrix object in R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">G</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">genotype_matrix</span><span class="p">()</span>
<span class="n">G</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 13 × 8 of type int</caption>
<tbody>
	<tr><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>1</td></tr>
	<tr><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>1</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>We can then use R functions directly on the genotype matrix:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">allele_frequency</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rowMeans</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
<span class="n">allele_frequency</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>0.125</li><li>0.125</li><li>0.125</li><li>0.125</li><li>0.5</li><li>0.625</li><li>0.875</li><li>0.875</li><li>0.125</li><li>0.25</li><li>0.125</li><li>0.625</li><li>0.125</li></ol>
</div></div>
</div>
</section>
<section id="jupyter-notebook-tips">
<h2>Jupyter notebook tips<a class="headerlink" href="#jupyter-notebook-tips" title="Permalink to this headline">#</a></h2>
<p>When running R within a <a class="reference external" href="https://jupyter.org">Jupyter notebook</a>, a few magic functions
can be defined that allow tskit objects to be rendered within the notebook:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define some magic functions to allow objects to be displayed in R Jupyter notebooks</span>
<span class="n">repr_html.tskit.trees.TreeSequence</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`_repr_html_`</span><span class="p">()}</span>
<span class="n">repr_html.tskit.trees.Tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`_repr_html_`</span><span class="p">()}</span>
<span class="n">repr_svg.tskit.drawing.SVGString</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="kc">...</span><span class="p">){</span><span class="n">obj</span><span class="o">$</span><span class="nf">`__str__`</span><span class="p">()}</span>
</pre></div>
</div>
</div>
</div>
<p>This leads to much nicer tabular summaries:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div>
  <style>
    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
    .tskit-table tbody tr td {padding: 0.5em 0.5em;}
    .tskit-table tbody tr td:first-of-type {text-align: left;}
    .tskit-details-label {vertical-align: top; padding-right:5px;}
    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
    .tskit-table-set-table {margin: 12px 0 0 12px;}
    details {display: inline-block;}
    summary {cursor: pointer; outline: 0; display: list-item;}
  </style>
  <div class="tskit-table-set">
    <div class="tskit-table-set-table">
      <table class="tskit-table">
        <thead>
          <tr>
            <th style="padding:0;line-height:21px;">
              <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
              <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr><td>Trees</td><td>6</td></tr>
          <tr><td>Sequence Length</td><td>6000.0</td></tr>
          <tr><td>Time Units</td><td>generations</td></tr>
          <tr><td>Sample Nodes</td><td>8</td></tr>
          <tr><td>Total Size</td><td>6.0 KiB</td></tr>
          <tr>
            <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
        </tbody>
      </table>
    </div>
    <div class="tskit-table-set-table">
      <table class="tskit-table">
        <thead>
          <tr>
            <th style="line-height:21px;">Table</th>
            <th>Rows</th>
            <th>Size</th>
            <th>Has Metadata</th>
          </tr>
        </thead>
        <tbody>

      <tr>
        <td>Edges</td>
          <td>32</td>
          <td>1.0 KiB</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Individuals</td>
          <td>4</td>
          <td>136 Bytes</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Migrations</td>
          <td>0</td>
          <td>8 Bytes</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Mutations</td>
          <td>13</td>
          <td>497 Bytes</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Nodes</td>
          <td>20</td>
          <td>568 Bytes</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Populations</td>
          <td>1</td>
          <td>224 Bytes</td>
          <td style="text-align: center;">
            ✅
          </td>
        </tr>

      <tr>
        <td>Provenances</td>
          <td>5</td>
          <td>3.0 KiB</td>
          <td style="text-align: center;">

          </td>
        </tr>

      <tr>
        <td>Sites</td>
          <td>13</td>
          <td>341 Bytes</td>
          <td style="text-align: center;">

          </td>
        </tr>

        </tbody>
      </table>
    </div>
  </div>
</div>
</div></div>
</div>
<p>It also allows trees and tree sequences to be plotted inline:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">y_ticks</span><span class="o">=</span><span class="m">0</span><span class="o">:</span><span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tskitr_23_0.svg" src="_images/tskitr_23_0.svg" /></div>
</div>
</section>
<section id="interaction-with-r-libraries">
<h2>Interaction with R libraries<a class="headerlink" href="#interaction-with-r-libraries" title="Permalink to this headline">#</a></h2>
<p>R has a number of libraries to deal with genomic data and trees. Below we focus on the
phylogenetic tree representation defined in the the popular
<a class="reference external" href="http://ape-package.ird.fr">ape</a> package, taking all the trees
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.write_nexus" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">exported</span> <span class="pre">in</span> <span class="pre">Nexus</span> <span class="pre">format</span></code></a>, or
individual trees <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.as_newick" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">exported</span> <span class="pre">in</span> <span class="pre">Newick</span> <span class="pre">format</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">tempfile</span><span class="p">()</span>
<span class="n">ts_mut</span><span class="o">$</span><span class="nf">write_nexus</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="c1"># Warning - ape trees are stored independently, so this will use much more memory than tskit</span>
<span class="n">trees</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ape</span><span class="o">::</span><span class="nf">read.nexus</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="n">force.multi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">  </span><span class="c1"># return a set of trees</span>

<span class="c1"># Or simply read in a single tree</span>
<span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ape</span><span class="o">::</span><span class="nf">read.tree</span><span class="p">(</span><span class="n">text</span><span class="o">=</span><span class="n">ts_mut</span><span class="o">$</span><span class="nf">first</span><span class="p">()</span><span class="o">$</span><span class="nf">as_newick</span><span class="p">())</span>

<span class="c1"># Now we can plot the tree in tskit style, but using the ape library</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">direction</span><span class="o">=</span><span class="s">&quot;downward&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">srt</span><span class="o">=</span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">adj</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span><span class="w">  </span><span class="c1"># or equivalently use trees[[1]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/tskitr_25_0.png" src="_images/tskitr_25_0.png" />
</div>
</div>
<p>Note that nodes are labelled with the prefix <code class="docutils literal notranslate"><span class="pre">n</span></code>, so that nodes <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, …
become <code class="docutils literal notranslate"><span class="pre">n0</span></code>, <code class="docutils literal notranslate"><span class="pre">n1</span></code>, <code class="docutils literal notranslate"><span class="pre">n2</span></code> … etc. This helps to avoid
confusion between the the zero-based counting system used natively
by <code class="docutils literal notranslate"><span class="pre">tskit</span></code>, and the one-based counting system used in <code class="docutils literal notranslate"><span class="pre">R</span></code>.</p>
</section>
<section id="further-information">
<h2>Further information<a class="headerlink" href="#further-information" title="Permalink to this headline">#</a></h2>
<p>Be sure to check out the <a class="reference external" href="https://rstudio.github.io/reticulate/">reticulate</a>
documentation, in particular on
<a class="reference external" href="https://rstudio.github.io/reticulate/articles/calling_python.html">Calling Python from R</a>,
which includes important information on how R data types are converted to their
equivalent Python types.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="forward_sims.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><em>Building a forward simulator</em></p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="tutorial_development.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><em>Developing new tutorials</em></p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Tskit Developers<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>